// @evershop/evershop/dist/components/frontStore/cart/CartTotalSummary.js
import Area from '@components/common/Area.js';
import { useAppState } from '@components/common/context/app.js';
import { useCartState } from '@components/frontStore/cart/CartContext.js';
import { CouponForm } from '@components/frontStore/CouponForm.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
const SkeletonValue = ({ children, loading = false, className = '' }) => {
    if (!loading) {
        return React.createElement(React.Fragment, null, children);
    }
    return (React.createElement("span", { className: `relative ${className}` },
        React.createElement("span", { className: "opacity-0" }, children),
        React.createElement("span", { className: "absolute inset-0 bg-gray-200 rounded animate-pulse" })));
};
const Total = ({ total, totalTaxAmount, priceIncludingTax, loading = false }) => {
    return (React.createElement("div", { className: "summary__row grand-total flex justify-between py-2" },
        (priceIncludingTax && (React.createElement("div", null,
            React.createElement("div", { className: "font-bold" },
                React.createElement("span", null, _('Total'))),
            React.createElement("div", null,
                React.createElement("span", { className: "italic font-normal" },
                    "(",
                    _('Inclusive of tax ${totalTaxAmount}', { totalTaxAmount }),
                    ")"))))) || React.createElement("span", { className: "self-center font-bold" }, _('Total')),
        React.createElement("div", null,
            React.createElement("div", null),
            React.createElement(SkeletonValue, { loading: loading, className: "grand-total-value" }, total))));
};
const Tax = ({ showPriceIncludingTax, amount, loading = false }) => {
    if (showPriceIncludingTax) {
        return null;
    }
    return (React.createElement("div", { className: "summary-row flex justify-between py-2" },
        React.createElement("span", null, _('Tax')),
        React.createElement("div", null,
            React.createElement("div", null),
            React.createElement(SkeletonValue, { loading: loading, className: "text-right" }, amount))));
};
const Subtotal = ({ subTotal, loading = false }) => {
    return (React.createElement("div", { className: "flex justify-between gap-7 py-2" },
        React.createElement("div", null, _('Sub total')),
        React.createElement(SkeletonValue, { loading: loading, className: "text-right" }, subTotal)));
};
const Discount = ({ discountAmount, coupon, loading = false }) => {
    if (!coupon) {
        return (React.createElement("div", { className: "gap-7 py-2" },
            React.createElement(CouponForm, null)));
    }
    return (React.createElement("div", { className: "flex justify-between gap-7 py-2" },
        React.createElement("div", null, _('Discount(${coupon})', { coupon })),
        React.createElement(SkeletonValue, { loading: loading, className: "text-right" }, discountAmount)));
};
const Shipping = ({ method, cost, loading = false }) => {
    return (React.createElement("div", { className: "summary-row flex justify-between gap-7 py-2" },
        method && (React.createElement(React.Fragment, null,
            React.createElement("span", null, _('Shipping (${method})', { method })),
            React.createElement("div", null,
                React.createElement(SkeletonValue, { loading: loading }, cost)))),
        !method && (React.createElement(React.Fragment, null,
            React.createElement("span", null, _('Shipping')),
            React.createElement("span", { className: "text-gray-500 italic font-normal" }, _('Select shipping method'))))));
};
const DefaultCartSummary = ({ loading, showPriceIncludingTax, subTotal, discountAmount, coupon, shippingMethod, shippingCost, taxAmount, total }) => (React.createElement("div", { className: "cart__total__summary font-semibold" },
    React.createElement(Area, { id: "cartSummaryBeforeSubTotal", noOuter: true }),
    React.createElement(Subtotal, { subTotal: subTotal, loading: loading }),
    React.createElement(Area, { id: "cartSummaryAfterSubTotal", noOuter: true }),
    React.createElement(Area, { id: "cartSummaryBeforeDiscount", noOuter: true }),
    React.createElement(Discount, { discountAmount: discountAmount, coupon: coupon, loading: loading }),
    React.createElement(Area, { id: "cartSummaryAfterDiscount", noOuter: true }),
    React.createElement(Area, { id: "cartSummaryBeforeShipping", noOuter: true }),
    React.createElement(Shipping, { method: shippingMethod, cost: shippingCost, loading: loading }),
    React.createElement(Area, { id: "cartSummaryAfterShipping", noOuter: true }),
    React.createElement(Area, { id: "cartSummaryBeforeTax", noOuter: true }),
    React.createElement(Tax, { amount: taxAmount, showPriceIncludingTax: showPriceIncludingTax, loading: loading }),
    React.createElement(Area, { id: "cartSummaryAfterTax", noOuter: true }),
    React.createElement(Area, { id: "cartSummaryBeforeTotal", noOuter: true }),
    React.createElement(Total, { total: total, totalTaxAmount: taxAmount, priceIncludingTax: showPriceIncludingTax, loading: loading }),
    React.createElement(Area, { id: "cartSummaryAfterTotal", noOuter: true })));
function CartTotalSummary({ children }) {
    var _a, _b, _c, _d, _e, _f, _g;
    const { data: cart, loadingStates } = useCartState();
    const { config: { tax: { priceIncludingTax } } } = useAppState();
    const subTotal = priceIncludingTax
        ? ((_a = cart === null || cart === void 0 ? void 0 : cart.subTotalInclTax) === null || _a === void 0 ? void 0 : _a.text) || ''
        : ((_b = cart === null || cart === void 0 ? void 0 : cart.subTotal) === null || _b === void 0 ? void 0 : _b.text) || '';
    const discountAmount = ((_c = cart === null || cart === void 0 ? void 0 : cart.discountAmount) === null || _c === void 0 ? void 0 : _c.text) || '';
    const coupon = cart === null || cart === void 0 ? void 0 : cart.coupon;
    const shippingMethod = cart === null || cart === void 0 ? void 0 : cart.shippingMethodName;
    const shippingCost = priceIncludingTax
        ? ((_d = cart === null || cart === void 0 ? void 0 : cart.shippingFeeInclTax) === null || _d === void 0 ? void 0 : _d.text) || ''
        : ((_e = cart === null || cart === void 0 ? void 0 : cart.shippingFeeExclTax) === null || _e === void 0 ? void 0 : _e.text) || '';
    const taxAmount = ((_f = cart === null || cart === void 0 ? void 0 : cart.totalTaxAmount) === null || _f === void 0 ? void 0 : _f.text) || '';
    const total = ((_g = cart === null || cart === void 0 ? void 0 : cart.grandTotal) === null || _g === void 0 ? void 0 : _g.text) || '';
    return (React.createElement("div", { className: "grid grid-cols-1 gap-5" }, children ? (children({
        loading: Object.values(loadingStates).some((state) => state === true || (typeof state === 'string' && state !== null)),
        showPriceIncludingTax: priceIncludingTax,
        subTotal,
        discountAmount,
        coupon,
        shippingMethod,
        shippingCost,
        taxAmount,
        total
    })) : (React.createElement(DefaultCartSummary, { loading: Object.values(loadingStates).some((state) => state === true || (typeof state === 'string' && state !== null)), showPriceIncludingTax: priceIncludingTax, subTotal: subTotal, discountAmount: discountAmount, coupon: coupon, shippingMethod: shippingMethod, shippingCost: shippingCost, taxAmount: taxAmount, total: total }))));
}
export { CartTotalSummary, DefaultCartSummary, Subtotal, Discount, Shipping, Tax, Total };
