// @evershop/evershop/dist/components/frontStore/cart/DefaultCartItemList.js
import { Area } from '@components/common/Area.js';
import { ExtendableTable } from '@components/common/ExtendableTable.js';
import { Image } from '@components/common/Image.js';
import { ProductNoThumbnail } from '@components/common/ProductNoThumbnail.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
import { ItemQuantity } from './ItemQuantity.js';
export const DefaultCartItemList = ({ items, showPriceIncludingTax = true, loading = false, onSort, currentSort, onRemoveItem }) => {
    const columns = [
        {
            key: 'productInfo',
            header: { label: _('Product'), className: '' },
            className: 'font-medium align-top',
            sortable: false,
            render: (row) => {
                var _a, _b, _c, _d;
                const priceValue = showPriceIncludingTax
                    ? (_a = row.productPriceInclTax) === null || _a === void 0 ? void 0 : _a.text
                    : (_b = row.productPrice) === null || _b === void 0 ? void 0 : _b.text;
                return (React.createElement("div", { className: "flex justify-start gap-4" },
                    React.createElement("div", null, row.thumbnail ? (React.createElement(Image, { src: row.thumbnail, alt: row.productName, width: 80, height: 80, className: "rounded-md" })) : (React.createElement(ProductNoThumbnail, { width: 80, height: 80 }))),
                    React.createElement("div", { className: "font-medium flex flex-col gap-1 items-start h-full" },
                        React.createElement("span", { className: "font-semibold" }, row.productName), (_c = row.variantOptions) === null || _c === void 0 ? void 0 :
                        _c.map((option) => (React.createElement("span", { key: option.optionId, className: "text-xs text-muted" },
                            option.attributeName,
                            ": ",
                            option.optionText))),
                        React.createElement("span", { className: "text-sm text-muted" },
                            priceValue,
                            " x ",
                            row.qty),
                        React.createElement("a", { href: "#", className: "text-red-500 text-sm", onClick: (e) => {
                                e.preventDefault();
                                onRemoveItem === null || onRemoveItem === void 0 ? void 0 : onRemoveItem(row.cartItemId);
                            } }, _('Remove')), (_d = row.errors) === null || _d === void 0 ? void 0 :
                        _d.map((error, index) => (React.createElement("span", { key: index, className: "text-xs text-red-500" }, error))))));
            }
        },
        {
            key: 'qty',
            header: { label: _('Quantity'), className: 'text-center' },
            sortable: true,
            render: (row) => {
                return (React.createElement("div", { className: "text-left" },
                    React.createElement(ItemQuantity, { initialValue: row.qty, cartItemId: row.cartItemId, min: 1, max: 99 }, ({ quantity, increase, decrease }) => (React.createElement("div", { className: "flex items-center" },
                        React.createElement("button", { onClick: decrease, disabled: loading || quantity <= 1, className: "px-1 disabled:opacity-50 text-lg" }, "\u2212"),
                        React.createElement("span", { className: "min-w-[3rem] text-center" }, quantity),
                        React.createElement("button", { onClick: increase, disabled: loading, className: "disabled:opacity-50 text-lg" }, "+"))))));
            }
        },
        {
            key: 'lineTotal',
            header: { label: _('Total'), className: 'flex justify-end' },
            sortable: true,
            render: (row) => {
                var _a, _b;
                const totalValue = showPriceIncludingTax
                    ? (_a = row.lineTotalInclTax) === null || _a === void 0 ? void 0 : _a.text
                    : (_b = row.lineTotal) === null || _b === void 0 ? void 0 : _b.text;
                return (React.createElement("div", { className: "text-right" },
                    React.createElement("span", { className: "font-bold" }, totalValue)));
            }
        }
    ];
    const [rows, setRows] = React.useState(items);
    React.useEffect(() => {
        setRows(items);
    }, [items]);
    return (React.createElement(React.Fragment, null,
        React.createElement(Area, { id: "miniCartItemListBefore", noOuter: true }),
        React.createElement(ExtendableTable, { name: "shoppingCartItems", columns: columns, initialData: rows, loading: loading, emptyMessage: _('Your cart is empty'), onSort: onSort, currentSort: currentSort, className: "cart__items__table border-none table-auto border-spacing-y-2 border-separate w-full" }),
        React.createElement(Area, { id: "miniCartItemListAfter", noOuter: true }),
        React.createElement("style", null, `
        .cart__items__table th, .cart__items__table td {
          padding: 0.75rem;
        }
        .cart__items__table th {
          border: none;
        }
        .cart__items__table td {
          border: none;
        }
      `)));
};
