// @evershop/evershop/dist/modules/setting/pages/admin/paymentSetting/PaymentSetting.js
import { SettingMenu } from '@components/admin/SettingMenu.js';
import Area from '@components/common/Area.js';
import { Form } from '@components/common/form/Form.js';
import React from 'react';
export default function PaymentSetting({ saveSettingApi }) {
    return (React.createElement("div", { className: "main-content-inner" },
        React.createElement("div", { className: "grid grid-cols-6 gap-x-5 grid-flow-row " },
            React.createElement("div", { className: "col-span-2" },
                React.createElement(SettingMenu, null)),
            React.createElement("div", { className: "col-span-4" },
                React.createElement(Form, { id: "paymentSettingForm", method: "POST", action: saveSettingApi, successMessage: "Payment setting saved" },
                    React.createElement(Area, { id: "paymentSetting", className: "grid gap-5" }))))));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    saveSettingApi: url(routeId: "saveSetting")
  }
`;
