// @evershop/evershop/dist/modules/promotion/pages/admin/couponEdit+couponNew/components/BuyXGetY.js
import { ProductSelector } from '@components/admin/ProductSelector.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import React, { useEffect } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
const SkuSelector = ({ product, updateProduct }) => {
    const modal = useModal();
    const onSelect = (sku) => {
        updateProduct({
            ...product,
            sku
        });
        modal.close();
    };
    return (React.createElement("div", null,
        React.createElement("a", { href: "#", className: "text-interactive hover:underline", onClick: (e) => {
                e.preventDefault();
                modal.open();
            } }, product.sku ? (React.createElement("span", { className: "italic" },
            "\u2018",
            product.sku,
            "\u2019")) : (React.createElement("span", null, "Choose SKU"))),
        React.createElement(Modal, { title: "Select SKU", onClose: modal.close, isOpen: modal.isOpen },
            React.createElement(ProductSelector, { selectedProducts: [product].map((p) => ({
                    sku: p.sku,
                    uuid: undefined,
                    productId: undefined
                })), onSelect: onSelect, onUnSelect: () => { } }))));
};
const BuyXGetYList = ({ requireProducts }) => {
    const { unregister } = useFormContext();
    const { fields, append, remove, update, replace } = useFieldArray({
        name: 'buyx_gety'
    });
    useEffect(() => {
        replace(requireProducts.map((product) => ({
            sku: product.sku,
            buyQty: typeof product.buyQty === 'string'
                ? parseInt(product.buyQty) || 1
                : product.buyQty,
            getQty: typeof product.getQty === 'string'
                ? parseInt(product.getQty) || 1
                : product.getQty,
            maxY: typeof product.maxY === 'string'
                ? parseInt(product.maxY) || 2
                : product.maxY,
            discount: typeof product.discount === 'string'
                ? parseInt(product.discount) || 100
                : product.discount
        })));
        return () => {
            unregister('buyx_gety'); // Cleanup: unregister field when component unmounts
        };
    }, []);
    return (React.createElement("div", null,
        React.createElement("table", { className: "table table-bordered" },
            React.createElement("thead", null,
                React.createElement("tr", null,
                    React.createElement("th", null,
                        React.createElement("span", null, "Sku")),
                    React.createElement("th", null,
                        React.createElement("span", null, "X")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Y")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Max of Y")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Discount percent")),
                    React.createElement("th", null, " "))),
            React.createElement("tbody", null, fields.map((p, i) => (React.createElement("tr", { key: p.id },
                React.createElement("td", null,
                    React.createElement(SkuSelector, { product: p, updateProduct: (product) => {
                            update(i, {
                                ...p,
                                sku: product.sku
                            });
                        } })),
                React.createElement("td", null,
                    React.createElement(NumberField, { name: `buyx_gety.${i}.buy_qty`, defaultValue: p.buyQty, placeholder: "Buy qty", required: true, validation: {
                            required: 'Buy qty is required'
                        } })),
                React.createElement("td", null,
                    React.createElement(NumberField, { name: `buyx_gety.${i}.get_qty`, defaultValue: p.getQty, placeholder: "Get qty", required: true, validation: {
                            required: 'Get qty is required'
                        } })),
                React.createElement("td", null,
                    React.createElement(NumberField, { name: `buyx_gety.${i}.max_y`, defaultValue: p.maxY, placeholder: "Max of Y", required: true, validation: {
                            required: 'Max of Y is required'
                        } })),
                React.createElement("td", null,
                    React.createElement(NumberField, { name: `buyx_gety.${i}.discount`, defaultValue: p.discount, placeholder: "Discount percent", required: true, validation: {
                            required: 'Discount percent is required'
                        }, unit: "%" })),
                React.createElement("td", null,
                    React.createElement("a", { className: "text-critical", href: "#", onClick: (e) => {
                            e.preventDefault();
                            remove(i);
                        } },
                        React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                            React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M18 12H6" }))))))))),
        React.createElement("div", { className: "mt-2 flex justify-start" },
            React.createElement("div", { className: "items-center flex" },
                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M12 6v6m0 0v6m0-6h6m-6 0H6" }))),
            React.createElement("div", { className: "pl-2" },
                React.createElement("a", { href: "#", onClick: (e) => {
                        e.preventDefault();
                        append({
                            sku: '',
                            buyQty: 1,
                            getQty: 1,
                            maxY: 2,
                            discount: 100
                        });
                    } },
                    React.createElement("span", null, "Add product"))))));
};
const BuyXGetY = ({ requireProducts }) => {
    const { watch } = useFormContext();
    const watchDiscountType = watch('discount_type');
    if (watchDiscountType !== 'buy_x_get_y') {
        return null;
    }
    return React.createElement(BuyXGetYList, { requireProducts: requireProducts });
};
export { BuyXGetY };
