// @evershop/evershop/dist/modules/promotion/pages/admin/couponEdit+couponNew/components/conditionCriterias.js
export var Operator;
(function (Operator) {
    Operator["EQUAL"] = "=";
    Operator["NOT_EQUAL"] = "!=";
    Operator["GREATER"] = ">";
    Operator["GREATER_OR_EQUAL"] = ">=";
    Operator["SMALLER"] = "<";
    Operator["SMALLER_OR_EQUAL"] = "<=";
    Operator["IN"] = "IN";
    Operator["NOT_IN"] = "NOT IN";
})(Operator || (Operator = {}));
const options = [
    {
        key: 'category',
        label: 'Category'
    },
    {
        key: 'collection',
        label: 'Collection'
    },
    {
        key: 'attribute_group',
        label: 'Attribute Group'
    },
    {
        key: 'sku',
        label: 'Sku'
    },
    {
        key: 'price',
        label: 'Price'
    }
];
const operators = [
    {
        key: Operator.EQUAL,
        label: 'Equal'
    },
    {
        key: Operator.NOT_EQUAL,
        label: 'Not equal'
    },
    {
        key: Operator.GREATER,
        label: 'Greater'
    },
    {
        key: Operator.GREATER_OR_EQUAL,
        label: 'Greater or equal'
    },
    {
        key: Operator.SMALLER,
        label: 'Smaller'
    },
    {
        key: Operator.SMALLER_OR_EQUAL,
        label: 'Equal or smaller'
    },
    {
        key: Operator.IN,
        label: 'In'
    },
    {
        key: Operator.NOT_IN,
        label: 'Not in'
    }
];
export { options, operators };
