// @evershop/evershop/dist/modules/promotion/pages/admin/couponEdit+couponNew/components/RequireProducts.js
import { InputField } from '@components/common/form/InputField.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import React, { useEffect } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
import { options, operators, Operator } from './conditionCriterias.js';
import { ValueSelector } from './ValueSelector.js';
export function RequiredProducts({ requiredProducts }) {
    const { setValue, watch } = useFormContext();
    const { fields, append, remove, replace } = useFieldArray({
        name: 'condition.required_products'
    });
    useEffect(() => {
        replace(requiredProducts);
    }, []);
    const fieldsWatch = watch('condition.required_products');
    return (React.createElement("div", { style: { marginTop: '1rem', marginBottom: '1rem' } },
        React.createElement("div", null,
            React.createElement("span", null, "Order must contains product matched bellow conditions(All)")),
        React.createElement("table", { className: "table table-auto", style: { marginTop: 0 } },
            React.createElement("thead", null,
                React.createElement("tr", null,
                    React.createElement("th", null,
                        React.createElement("span", null, "Key")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Operator")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Value")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Minimum quantity")),
                    React.createElement("th", null, " "))),
            React.createElement("tbody", null, fields.map((p, i) => {
                var _a, _b;
                return (React.createElement("tr", { key: p.id },
                    React.createElement("td", null, p.editable ? (React.createElement(SelectField, { name: `condition.required_products.${i}.key`, defaultValue: p.key, options: options.map((option) => ({
                            value: option.key,
                            label: option.label
                        })), wrapperClassName: "form-field mb-0" })) : (React.createElement(React.Fragment, null,
                        React.createElement(InputField, { type: "hidden", name: `condition.required_products.${i}.key`, readOnly: true, value: p.key, wrapperClassName: "form-field mb-0" }),
                        React.createElement(InputField, { name: `condition.required_products.${i}.keylabel`, readOnly: true, value: ((_a = options.find((c) => c.key === p.key)) === null || _a === void 0 ? void 0 : _a.label) || 'Unknown', wrapperClassName: "form-field mb-0" })))),
                    React.createElement("td", null, p.editable ? (React.createElement(SelectField, { options: operators.map((operator) => ({
                            value: operator.key,
                            label: operator.label
                        })), name: `condition.required_products.${i}.operator`, defaultValue: p.operator, wrapperClassName: "form-field mb-0" })) : (React.createElement(React.Fragment, null,
                        React.createElement(InputField, { type: "hidden", name: `condition.required_products.${i}.operator`, readOnly: true, value: p.operator, wrapperClassName: "form-field mb-0" }),
                        React.createElement(InputField, { readOnly: true, name: `condition.required_products.${i}.operatorlabel`, value: ((_b = operators.find((c) => c.key === p.operator)) === null || _b === void 0 ? void 0 : _b.label) ||
                                'Unknown', wrapperClassName: "form-field mb-0" })))),
                    React.createElement("td", null,
                        fieldsWatch[i].key === 'price' && (React.createElement(NumberField, { name: `condition.required_products.${i}.value`, defaultValue: p.value, wrapperClassName: "form-field mb-0" })),
                        fieldsWatch[i].key !== 'price' && (React.createElement(React.Fragment, null,
                            React.createElement(InputField, { type: "hidden", name: `condition.required_products.${i}.value`, value: p.value, wrapperClassName: "form-field mb-0" }),
                            React.createElement(ValueSelector, { condition: fieldsWatch[i], updateCondition: (values) => {
                                    setValue(`condition.required_products.${i}.value`, values);
                                } })))),
                    React.createElement("td", null,
                        React.createElement("div", { style: { width: '80px' } },
                            React.createElement(NumberField, { name: `condition.required_products.${i}.qty`, defaultValue: typeof p.qty === 'number'
                                    ? p.qty
                                    : parseInt(p.qty, 10) || 1, placeholder: "Enter the quantity", required: true, validation: {
                                    required: 'Minimum quantity is required',
                                    min: {
                                        value: 1,
                                        message: ''
                                    }
                                }, wrapperClassName: "form-field mb-0" }))),
                    React.createElement("td", null,
                        React.createElement("a", { href: "#", className: "text-critical", onClick: (e) => {
                                e.preventDefault();
                                remove(i);
                            } },
                            React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                                React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M18 12H6" }))))));
            }))),
        React.createElement("div", { className: "mt-2 flex justify-start" },
            React.createElement("div", { className: "items-center flex" },
                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M12 6v6m0 0v6m0-6h6m-6 0H6" }))),
            React.createElement("div", { className: "pl-2" },
                React.createElement("a", { href: "#", onClick: (e) => {
                        e.preventDefault();
                        append({
                            key: 'category',
                            operator: Operator.EQUAL,
                            value: '',
                            qty: '',
                            editable: true
                        });
                    } },
                    React.createElement("span", null, "Add product"))))));
}
RequiredProducts.defaultProps = {
    requiredProducts: []
};
