// @evershop/evershop/dist/modules/promotion/pages/admin/couponEdit+couponNew/components/TargetProducts.js
import { InputField } from '@components/common/form/InputField.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import React, { useEffect } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
import { options, operators, Operator } from './conditionCriterias.js';
import { ValueSelector } from './ValueSelector.js';
function Products({ targetProducts, maxQty }) {
    const { setValue, watch, unregister } = useFormContext();
    const { fields, append, remove, replace } = useFieldArray({
        name: 'target_products.products'
    });
    const watchDiscountType = watch('discount_type');
    const fieldWatch = watch('target_products.products');
    useEffect(() => {
        replace(targetProducts.map((product) => ({
            key: product.key,
            operator: product.operator,
            value: product.value
        })));
        return () => {
            unregister('target_products.products');
        };
    }, []);
    if (watchDiscountType !== 'fixed_discount_to_specific_products' &&
        watchDiscountType !== 'percentage_discount_to_specific_products') {
        return null;
    }
    return (React.createElement("div", null,
        React.createElement("div", { className: "mb-2 mt-2" },
            React.createElement("div", { className: "flex justify-start items-center" },
                React.createElement("div", null, "Maximum"),
                React.createElement("div", { style: { width: '100px', padding: '0 1rem' } },
                    React.createElement(NumberField, { name: "target_products.maxQty", defaultValue: maxQty, placeholder: "10", required: true, validation: {
                            required: 'Maximum quantity is required',
                            min: {
                                value: 0,
                                message: 'Maximum quantity must be greater than or equal to 0'
                            }
                        }, min: 0, wrapperClassName: "form-field mb-0" })),
                React.createElement("div", null, "quantity of products are matched bellow conditions(All)"))),
        React.createElement("table", { className: "table table-bordered", style: { marginTop: 0 } },
            React.createElement("thead", null,
                React.createElement("tr", null,
                    React.createElement("th", null,
                        React.createElement("span", null, "Key")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Operator")),
                    React.createElement("th", null,
                        React.createElement("span", null, "Value")),
                    React.createElement("th", null, " "))),
            React.createElement("tbody", null, fields.map((product, index) => {
                var _a, _b;
                return (React.createElement("tr", { key: product.id },
                    React.createElement("td", null, product.editable ? (React.createElement(SelectField, { name: `target_products.products.${index}.key`, wrapperClassName: "form-field mb-0", defaultValue: product.key, disabled: !product.editable, options: options.map((option) => ({
                            value: option.key,
                            label: option.label
                        })) })) : (React.createElement(React.Fragment, null,
                        React.createElement(InputField, { type: "hidden", name: `target_products.products.${index}.key`, readOnly: true, wrapperClassName: "form-field mb-0", value: product.key }),
                        React.createElement(InputField, { name: `target_products.products.${index}.keylabel`, readOnly: true, wrapperClassName: "form-field mb-0", value: ((_a = options.find((c) => c.key === product.key)) === null || _a === void 0 ? void 0 : _a.label) ||
                                'Unknown' })))),
                    React.createElement("td", null, product.editable ? (React.createElement(SelectField, { name: `target_products.products.${index}.operator`, defaultValue: product.operator, options: operators.map((operator) => ({
                            value: operator.key,
                            label: operator.label
                        })), wrapperClassName: "form-field mb-0", placeholder: "Select operator" })) : (React.createElement(React.Fragment, null,
                        React.createElement(InputField, { type: "hidden", name: `target_products.products.${index}.operator`, readOnly: true, wrapperClassName: "form-field mb-0", value: product.operator }),
                        React.createElement(InputField, { name: `target_products.products.${index}.operatorlabel`, type: "text", readOnly: true, wrapperClassName: "form-field mb-0", value: ((_b = options.find((c) => c.key === product.operator)) === null || _b === void 0 ? void 0 : _b.label) || 'Unknown' })))),
                    React.createElement("td", null,
                        fieldWatch[index].key === 'price' && (React.createElement(NumberField, { name: `target_products.products.${index}.value`, defaultValue: product.value, wrapperClassName: "form-field mb-0" })),
                        fieldWatch[index].key !== 'price' && (React.createElement(React.Fragment, null,
                            React.createElement(InputField, { type: "hidden", name: `target_products.products.${index}.value`, value: product.value, wrapperClassName: "form-field mb-0" }),
                            React.createElement(ValueSelector, { condition: fieldWatch[index], updateCondition: (values) => {
                                    setValue(`target_products.products.${index}.value`, values);
                                } })))),
                    React.createElement("td", null,
                        React.createElement("a", { href: "#", className: "text-critical", onClick: (e) => {
                                e.preventDefault();
                                remove(index);
                            } },
                            React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                                React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M18 12H6" }))))));
            }))),
        React.createElement("div", { className: "mt-2 flex justify-start" },
            React.createElement("div", { className: "items-center flex" },
                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "1.5rem", height: "1.5rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor", strokeWidth: 2 },
                    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M12 6v6m0 0v6m0-6h6m-6 0H6" }))),
            React.createElement("div", { className: "pl-2" },
                React.createElement("a", { href: "#", onClick: (e) => {
                        e.preventDefault();
                        append({
                            key: 'category',
                            operator: Operator.EQUAL,
                            value: '',
                            editable: true
                        });
                    }, className: "" },
                    React.createElement("span", null, "Add product"))))));
}
export function TargetProducts({ products, maxQty }) {
    const { watch } = useFormContext();
    const watchDiscountType = watch('discount_type');
    if (watchDiscountType !== 'fixed_discount_to_specific_products' &&
        watchDiscountType !== 'percentage_discount_to_specific_products') {
        return null;
    }
    return (React.createElement("div", null,
        React.createElement("h3", { className: "card-title" }, "Target products"),
        React.createElement(Products, { targetProducts: products, maxQty: maxQty })));
}
