// @evershop/evershop/dist/components/admin/AttributeGroupSelector.js
import { Card } from '@components/admin/Card.js';
import { SimplePagination } from '@components/common/SimplePagination.js';
import { CheckIcon } from '@heroicons/react/24/outline';
import React from 'react';
import { useQuery } from 'urql';
import './AttributeGroupSelector.scss';
const SearchQuery = `
  query Query ($filters: [FilterInput!]) {
    attributeGroups(filters: $filters) {
      items {
        attributeGroupId
        uuid
        groupName
      }
      total
    }
  }
`;
const AttributeGroupListSkeleton = () => {
    const skeletonItems = Array(5).fill(0);
    return (React.createElement("div", { className: "attribute-group-list-skeleton" }, skeletonItems.map((_, index) => (React.createElement("div", { key: index, className: "attribute-group-skeleton-item border-b flex justify-between items-center" },
        React.createElement("div", { className: "flex items-center" },
            React.createElement("div", null,
                React.createElement("div", { className: "skeleton-title h-5 w-30 bg-gray-200 rounded skeleton-pulse mb-2" }),
                React.createElement("div", { className: "skeleton-id h-4 w-20 bg-gray-200 rounded skeleton-pulse" }))),
        React.createElement("div", { className: "select-button" },
            React.createElement("div", { className: "skeleton-button h-6 w-12 bg-gray-200 rounded skeleton-pulse" })))))));
};
const isAttributeGroupSelected = (attributeGroup, selectedAttributeGroups) => {
    return selectedAttributeGroups.some((selected) => ((selected === null || selected === void 0 ? void 0 : selected.attributeGroupId) &&
        selected.attributeGroupId === attributeGroup.attributeGroupId) ||
        ((selected === null || selected === void 0 ? void 0 : selected.uuid) && selected.uuid === attributeGroup.uuid));
};
const AttributeGroupSelector = ({ onSelect, onUnSelect, selectedAttributeGroups }) => {
    var _a, _b;
    const [internalSelectedAttributeGroups, setInternalSelectedAttributeGroups] = React.useState(selectedAttributeGroups || []);
    const [loading, setLoading] = React.useState(false);
    const limit = 10;
    const [inputValue, setInputValue] = React.useState('');
    const [page, setPage] = React.useState(1);
    const [result, reexecuteQuery] = useQuery({
        query: SearchQuery,
        variables: {
            filters: inputValue
                ? [
                    { key: 'name', operation: 'like', value: inputValue },
                    { key: 'page', operation: 'eq', value: page.toString() },
                    { key: 'limit', operation: 'eq', value: limit.toString() }
                ]
                : [
                    { key: 'limit', operation: 'eq', value: limit.toString() },
                    { key: 'page', operation: 'eq', value: page.toString() }
                ]
        },
        pause: true
    });
    React.useEffect(() => {
        reexecuteQuery({ requestPolicy: 'network-only' });
    }, [page]);
    React.useEffect(() => {
        const timer = setTimeout(() => {
            setLoading(false);
            if (inputValue !== '') {
                reexecuteQuery({ requestPolicy: 'network-only' });
            }
        }, 1500);
        return () => clearTimeout(timer);
    }, [inputValue]);
    const { data, fetching, error } = result;
    if (error) {
        return (React.createElement("p", { className: "text-critical" },
            "There was an error fetching attribute groups.",
            error.message));
    }
    return (React.createElement(Card, { title: "Select Attribute Groups" },
        React.createElement(Card.Session, null,
            React.createElement("div", null,
                React.createElement("div", { className: "border rounded border-divider mb-5" },
                    React.createElement("input", { type: "text", value: inputValue || '', placeholder: "Search attribute groups", onChange: (e) => {
                            setInputValue(e.target.value);
                            setLoading(true);
                        } })),
                (fetching || loading) && React.createElement(AttributeGroupListSkeleton, null),
                !fetching && data && (React.createElement("div", { className: "divide-y" },
                    data.attributeGroups.items.length === 0 && (React.createElement("div", { className: "p-2 border border-divider rounded flex justify-center items-center" }, inputValue ? (React.createElement("p", null,
                        "No attribute groups found for query \"",
                        inputValue,
                        "\u201D")) : (React.createElement("p", null, "You have no attribute groups to display")))),
                    data.attributeGroups.items.map((a) => (React.createElement("div", { key: a.uuid, className: "grid grid-cols-8 gap-5 py-2 border-divider items-center" },
                        React.createElement("div", { className: "col-span-5" },
                            React.createElement("h3", null, a.groupName)),
                        React.createElement("div", { className: "col-span-3 text-right" },
                            !isAttributeGroupSelected(a, internalSelectedAttributeGroups) && (React.createElement("button", { type: "button", className: "button secondary", onClick: async (e) => {
                                    e.preventDefault();
                                    setInternalSelectedAttributeGroups((prev) => [
                                        ...prev,
                                        {
                                            attributeGroupId: a.attributeGroupId,
                                            uuid: a.uuid,
                                            groupName: a.groupName
                                        }
                                    ]);
                                    onSelect(a.attributeGroupId, a.uuid, a.groupName);
                                } }, "Select")),
                            isAttributeGroupSelected(a, internalSelectedAttributeGroups) && (React.createElement("a", { className: "button primary", href: "#", onClick: (e) => {
                                    e.preventDefault();
                                    setInternalSelectedAttributeGroups((prev) => prev.filter((c) => c.attributeGroupId !== a.attributeGroupId &&
                                        c.uuid !== a.uuid));
                                    onUnSelect(a.attributeGroupId, a.uuid, a.groupName);
                                } },
                                React.createElement(CheckIcon, { className: "w-5 h-5" }))))))))))),
        React.createElement(Card.Session, null,
            React.createElement("div", { className: "flex justify-between gap-5" },
                React.createElement(SimplePagination, { total: (data === null || data === void 0 ? void 0 : data.attributeGroups.total) || 0, count: ((_b = (_a = data === null || data === void 0 ? void 0 : data.attributeGroups) === null || _a === void 0 ? void 0 : _a.items) === null || _b === void 0 ? void 0 : _b.length) || 0, page: page, hasNext: limit * page < (data === null || data === void 0 ? void 0 : data.attributeGroups.total), setPage: setPage })))));
};
export { AttributeGroupSelector };
