// @evershop/evershop/dist/components/admin/CategoryTree.js
import { ChevronDownIcon } from '@heroicons/react/24/outline';
import { ChevronRightIcon } from '@heroicons/react/24/outline';
import { FolderIcon as Folder } from '@heroicons/react/24/outline';
import { MinusIcon } from '@heroicons/react/24/outline';
import React from 'react';
import { useQuery } from 'urql';
import './CategoryTree.scss';
import RenderIfTrue from '@components/common/RenderIfTrue.jsx';
const categoriesQuery = `
  query Query ($filters: [FilterInput]) {
    categories (filters: $filters) {
      items {
        categoryId,
        name
        hasChildren
        path {
          name
        }
      }
    }
  }
`;
const childrenQuery = `
  query Query ($filters: [FilterInput]) {
    categories (filters: $filters) {
      items {
        categoryId,
        name
        path {
          name
        }
        hasChildren
      }
    }
  }
`;
const Skeleton = () => (React.createElement("ul", { className: "skeleton-wrapper-category-tree" },
    React.createElement("li", { className: "skeleton mt-2" }),
    React.createElement("li", { className: "skeleton mt-2" }),
    React.createElement("li", { className: "skeleton mt-2" }),
    React.createElement("li", { className: "skeleton mt-2" })));
function CategoryItem({ category, selectedCategories, onSelect }) {
    const [expanded, setExpanded] = React.useState(false);
    const [result] = useQuery({
        query: childrenQuery,
        variables: {
            filters: [{ key: 'parent', operation: 'eq', value: category.categoryId }]
        },
        pause: !expanded
    });
    const { data, fetching, error } = result;
    if (error) {
        return (React.createElement("li", { className: "text-critical" },
            React.createElement("span", null, error.message)));
    }
    const className = (selectedCategories === null || selectedCategories === void 0 ? void 0 : selectedCategories.find((item) => item.categoryId === category.categoryId))
        ? 'flex justify-start gap-2 items-center p-2 rounded-md bg-green-100 transition-colors duration-500'
        : 'flex justify-start gap-2 items-center p-2 rounded-md hover:bg-gray-100 transition-colors duration-500';
    return (React.createElement("li", { className: "[&_ul]:pl-2" },
        React.createElement("div", { className: className },
            category.hasChildren && (React.createElement("a", { href: "#", onClick: (e) => {
                    e.preventDefault();
                    setExpanded(!expanded);
                } }, expanded ? (React.createElement(ChevronDownIcon, { width: 15, height: 15 })) : (React.createElement(ChevronRightIcon, { width: 15, height: 15 })))),
            !category.hasChildren && (React.createElement("span", { className: "text-gray-400" },
                React.createElement(MinusIcon, { width: 11, height: 15 }))),
            React.createElement("a", { href: "#", onClick: (e) => {
                    e.preventDefault();
                    onSelect(category);
                } },
                React.createElement("div", { className: "flex gap-2 justify-start items-center cursor-pointer" },
                    React.createElement(Folder, { width: 20, height: 20 }),
                    React.createElement("span", { className: "text-sm" }, category.name)))),
        data && data.categories.items.length > 0 && expanded && (React.createElement("div", { className: "pb-2" },
            React.createElement("ul", null, data.categories.items.map((child) => (React.createElement(CategoryItem, { key: child.value, category: child, selectedCategories: selectedCategories, onSelect: onSelect })))))),
        React.createElement(RenderIfTrue, { condition: fetching && expanded },
            React.createElement("div", { className: "pb-2" },
                React.createElement(Skeleton, null)))));
}
CategoryItem.defaultProps = {
    category: {},
    selectedCategory: {}
};
function CategoryTree({ selectedCategories, onSelect }) {
    const [result] = useQuery({
        query: categoriesQuery,
        variables: {
            filters: [{ key: 'parent', operation: 'eq', value: null }]
        }
    });
    const { data, fetching, error } = result;
    if (fetching) {
        return React.createElement(Skeleton, null);
    }
    if (error) {
        return React.createElement("p", { className: "text-critical" }, error.message);
    }
    if (!data || !data.categories || data.categories.items.length === 0) {
        return React.createElement("div", { className: "text-gray-400 text-md" }, "There is no category");
    }
    return (React.createElement("ul", { className: "category-tree" }, data.categories.items.map((category) => (React.createElement(CategoryItem, { key: category.value, category: category, selectedCategories: selectedCategories, onSelect: onSelect })))));
}
CategoryTree.defaultProps = {
    selectedCategories: []
};
export { CategoryTree };
