// @evershop/evershop/dist/components/admin/CategorySelector.js
import { SimplePagination } from '@components/common/SimplePagination.js';
import { CheckIcon } from '@heroicons/react/24/outline';
import React from 'react';
import { useQuery } from 'urql';
import './CategorySelector.scss';
const SearchQuery = `
  query Query ($filters: [FilterInput!]) {
    categories(filters: $filters) {
      items {
        categoryId
        uuid
        name
        path {
          name
        }
      }
      total
    }
  }
`;
const CategoryListSkeleton = () => {
    const skeletonItems = Array(5).fill(0);
    return (React.createElement("div", { className: "category-list-skeleton" }, skeletonItems.map((_, index) => (React.createElement("div", { key: index, className: "category-skeleton-item border-b flex justify-between items-center" },
        React.createElement("div", { className: "flex items-center" },
            React.createElement("div", null,
                React.createElement("div", { className: "skeleton-title h-5 w-30 bg-gray-200 rounded skeleton-pulse mb-2" }),
                React.createElement("div", { className: "skeleton-id h-4 w-20 bg-gray-200 rounded skeleton-pulse" }))),
        React.createElement("div", { className: "select-button" },
            React.createElement("div", { className: "skeleton-button h-6 w-12 bg-gray-200 rounded skeleton-pulse" })))))));
};
const isCategorySelected = (category, selectedCategories) => {
    return selectedCategories.some((selected) => ((selected === null || selected === void 0 ? void 0 : selected.categoryId) && selected.categoryId === category.categoryId) ||
        ((selected === null || selected === void 0 ? void 0 : selected.uuid) && selected.uuid === category.uuid));
};
const CategorySelector = ({ onSelect, onUnSelect, selectedCategories }) => {
    var _a, _b;
    const [internalSelectedCategories, setInternalSelectedCategories] = React.useState(selectedCategories || []);
    const [loading, setLoading] = React.useState(false);
    const limit = 10;
    const [inputValue, setInputValue] = React.useState('');
    const [page, setPage] = React.useState(1);
    const [result, reexecuteQuery] = useQuery({
        query: SearchQuery,
        variables: {
            filters: inputValue
                ? [
                    { key: 'name', operation: 'like', value: inputValue },
                    { key: 'page', operation: 'eq', value: page.toString() },
                    { key: 'limit', operation: 'eq', value: limit.toString() }
                ]
                : [
                    { key: 'limit', operation: 'eq', value: limit.toString() },
                    { key: 'page', operation: 'eq', value: page.toString() }
                ]
        },
        pause: true
    });
    React.useEffect(() => {
        reexecuteQuery({ requestPolicy: 'network-only' });
    }, [page]);
    React.useEffect(() => {
        const timer = setTimeout(() => {
            setLoading(false);
            if (inputValue !== '') {
                reexecuteQuery({ requestPolicy: 'network-only' });
            }
        }, 1500);
        return () => clearTimeout(timer);
    }, [inputValue]);
    const { data, fetching, error } = result;
    if (error) {
        return (React.createElement("p", { className: "text-critical" },
            "There was an error fetching categories.",
            error.message));
    }
    return (React.createElement("div", null,
        React.createElement("div", null,
            React.createElement("div", { className: "p-2" },
                React.createElement("div", { className: "form-field" },
                    React.createElement("input", { type: "text", value: inputValue || '', placeholder: "Search categories", onChange: (e) => {
                            setInputValue(e.target.value);
                            setLoading(true);
                        } }))),
            (fetching || loading) && React.createElement(CategoryListSkeleton, null),
            !fetching && data && (React.createElement("div", { className: "divide-y" },
                data.categories.items.length === 0 && (React.createElement("div", { className: "p-2 border border-divider rounded flex justify-center items-center" }, inputValue ? (React.createElement("p", null,
                    "No categories found for query \"",
                    inputValue,
                    "\u201D")) : (React.createElement("p", null, "You have no categories to display")))),
                data.categories.items.map((cat) => (React.createElement("div", { key: cat.uuid, className: "grid grid-cols-8 gap-5 py-2 border-divider items-center" },
                    React.createElement("div", { className: "col-span-5" },
                        React.createElement("h3", null, cat.path.map((item, index) => (React.createElement("span", { key: item.name, className: "text-gray-500" },
                            item.name,
                            index < cat.path.length - 1 && ' > '))))),
                    React.createElement("div", { className: "col-span-3 text-right" },
                        !isCategorySelected(cat, internalSelectedCategories) && (React.createElement("button", { type: "button", className: "button secondary", onClick: async (e) => {
                                e.preventDefault();
                                setInternalSelectedCategories((prev) => [
                                    ...prev,
                                    {
                                        categoryId: cat.categoryId,
                                        uuid: cat.uuid,
                                        name: cat.name
                                    }
                                ]);
                                onSelect(cat.categoryId, cat.uuid, cat.name);
                            } }, "Select")),
                        isCategorySelected(cat, internalSelectedCategories) && (React.createElement("a", { className: "button primary", href: "#", onClick: (e) => {
                                e.preventDefault();
                                setInternalSelectedCategories((prev) => prev.filter((c) => c.categoryId !== cat.categoryId &&
                                    c.uuid !== cat.uuid));
                                onUnSelect(cat.categoryId, cat.uuid, cat.name);
                            } },
                            React.createElement(CheckIcon, { className: "w-5 h-5" })))))))))),
        React.createElement("div", { className: "flex justify-between gap-5" },
            React.createElement(SimplePagination, { total: (data === null || data === void 0 ? void 0 : data.categories.total) || 0, count: ((_b = (_a = data === null || data === void 0 ? void 0 : data.categories) === null || _a === void 0 ? void 0 : _a.items) === null || _b === void 0 ? void 0 : _b.length) || 0, page: page, hasNext: limit * page < (data === null || data === void 0 ? void 0 : data.categories.total), setPage: setPage }))));
};
export { CategorySelector };
