// @evershop/evershop/dist/components/admin/ImageUploader.js
import React from 'react';
import { toast } from 'react-toastify';
import uniqid from 'uniqid';
import { useQuery } from 'urql';
import { get } from '../../lib/util/get.js';
import './ImageUploader.scss';
import Spinner from '@components/admin/Spinner.js';
import { ImageUploaderSkeleton } from './ImageUploaderSkeleton.js';
import { DndContext, closestCenter, KeyboardSensor, PointerSensor, useSensor, useSensors } from '@dnd-kit/core';
import { arrayMove, SortableContext, sortableKeyboardCoordinates, useSortable } from '@dnd-kit/sortable';
import { CSS } from '@dnd-kit/utilities';
const Upload = ({ imageUploadUrl, targetPath, onUpload, isSingleMode }) => {
    const [uploading, setUploading] = React.useState(false);
    const onChange = (e) => {
        setUploading(true);
        e.persist();
        const formData = new FormData();
        for (let i = 0; i < e.target.files.length; i += 1) {
            formData.append('images', e.target.files[i]);
        }
        formData.append('targetPath', targetPath || '');
        fetch(imageUploadUrl + (targetPath || ''), {
            method: 'POST',
            body: formData,
            headers: {
                'X-Requested-With': 'XMLHttpRequest'
            }
        })
            .then((response) => {
            const contentType = response.headers.get('content-type');
            if (!contentType || !contentType.includes('application/json')) {
                throw new TypeError('Something wrong. Please try again');
            }
            return response.json();
        })
            .then(async (response) => {
            if (!response.error) {
                await onUpload(get(response, 'data.files', []).map((i) => ({
                    uuid: uniqid(),
                    url: i.url,
                    path: i.path
                })));
            }
            else {
                toast.error(get(response, 'error.message', 'Failed!'));
            }
        })
            .catch((error) => {
            toast.error(error.message);
        })
            .finally(() => {
            e.target.value = null;
            setUploading(false);
        });
    };
    const id = uniqid();
    return (React.createElement("div", { className: "uploader grid-item" },
        React.createElement("div", { className: "uploader-icon" },
            React.createElement("label", { htmlFor: id }, uploading ? (React.createElement(Spinner, { width: isSingleMode ? 40 : 25, height: isSingleMode ? 40 : 25 })) : (React.createElement("svg", { style: {
                    width: isSingleMode ? '30px' : '30px',
                    height: isSingleMode ? '30px' : '30px'
                }, xmlns: "http://www.w3.org/2000/svg", className: "h-5 w-5", viewBox: "0 0 20 20", fill: "currentColor" },
                React.createElement("path", { fillRule: "evenodd", d: "M4 5a2 2 0 00-2 2v8a2 2 0 002 2h12a2 2 0 002-2V7a2 2 0 00-2-2h-1.586a1 1 0 01-.707-.293l-1.121-1.121A2 2 0 0011.172 3H8.828a2 2 0 00-1.414.586L6.293 4.707A1 1 0 015.586 5H4zm6 9a3 3 0 100-6 3 3 0 000 6z", clipRule: "evenodd" }))))),
        React.createElement("div", { className: "invisible" },
            React.createElement("input", { id: id, type: "file", multiple: true, onChange: onChange }))));
};
const Image = ({ image, allowDelete, onDelete, isFirst, isSingleMode }) => {
    const [deleting, setDeleting] = React.useState(false);
    // Use ref to track if component is mounted
    const isMounted = React.useRef(true);
    // Set up effect for cleanup
    React.useEffect(() => {
        return () => {
            // When component unmounts, set ref to false
            isMounted.current = false;
        };
    }, []);
    // Assign classes based on mode
    const classes = isSingleMode
        ? 'image'
        : `image grid-item ${isFirst ? 'first-item' : ''}`;
    return (React.createElement("div", { className: classes, id: image.uuid },
        React.createElement("div", { className: "img" },
            React.createElement("img", { src: image.url, alt: "" })),
        allowDelete && (React.createElement("span", { role: "button", tabIndex: 0, className: `remove cursor-pointer text-critical fill-current ${isSingleMode ? 'single-mode-remove' : ''}`, onClick: async () => {
                setDeleting(true);
                await onDelete(image);
                // Only update state if component is still mounted
                if (isMounted.current) {
                    setDeleting(false);
                }
            }, onKeyDown: () => { } },
            React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: isSingleMode ? '20' : '16', height: isSingleMode ? '20' : '16', viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round", className: "feather feather-trash-2" },
                React.createElement("polyline", { points: "3 6 5 6 21 6" }),
                React.createElement("path", { d: "M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2" }),
                React.createElement("line", { x1: "10", y1: "11", x2: "10", y2: "17" }),
                React.createElement("line", { x1: "14", y1: "11", x2: "14", y2: "17" })))),
        deleting && (React.createElement("div", { className: "remove" },
            React.createElement(Spinner, { width: 15, height: 15 })))));
};
const SortableImage = (props) => {
    const { attributes, listeners, setNodeRef, transform, transition } = useSortable({
        id: props.image.uuid
    });
    const style = {
        transform: CSS.Transform.toString(transform),
        transition
    };
    return (React.createElement("div", { ref: setNodeRef, style: style, className: `grid-item ${props.isFirst ? 'first-item' : ''}`, ...attributes, ...listeners },
        React.createElement(Image, { ...props })));
};
const GetUploadApiQuery = `
  query Query ($filters: [FilterInput!]) {
    imageUploadUrl: url(routeId: "imageUpload", params: [{key: "0", value: ""}])
  }
`;
const Images = ({ allowDelete = true, currentImages, imageUploadUrl, onDelete, onUpload, targetPath, isMultiple, allowSwap, onSortEnd }) => {
    const sensors = useSensors(useSensor(PointerSensor, {
        activationConstraint: {
            distance: 8
        }
    }), useSensor(KeyboardSensor, {
        coordinateGetter: sortableKeyboardCoordinates
    }));
    const handleDragEnd = (event) => {
        const { active, over } = event;
        if (active.id !== (over === null || over === void 0 ? void 0 : over.id) && onSortEnd && currentImages) {
            const oldIndex = currentImages.findIndex((img) => img.uuid === active.id);
            const newIndex = currentImages.findIndex((img) => img.uuid === (over === null || over === void 0 ? void 0 : over.id));
            if (oldIndex !== -1 && newIndex !== -1) {
                onSortEnd(oldIndex, newIndex);
            }
        }
    };
    if (!isMultiple) {
        const hasImage = currentImages && currentImages.length > 0;
        return (React.createElement("div", { className: `single-image-container ${!hasImage ? 'no-image' : ''}` },
            hasImage ? (React.createElement(Image, { key: currentImages[0].uuid, image: currentImages[0], onDelete: onDelete, allowDelete: allowDelete, isSingleMode: true })) : null,
            React.createElement(Upload, { imageUploadUrl: imageUploadUrl, targetPath: targetPath, onUpload: onUpload, isSingleMode: true })));
    }
    else if (allowSwap && currentImages && currentImages.length > 1) {
        return (React.createElement(DndContext, { sensors: sensors, collisionDetection: closestCenter, onDragEnd: handleDragEnd },
            React.createElement(SortableContext, { items: currentImages.map((img) => img.uuid) }, currentImages.map((image, index) => (React.createElement(SortableImage, { key: image.uuid, image: image, onDelete: onDelete, allowDelete: allowDelete, isFirst: index === 0 })))),
            React.createElement(Upload, { imageUploadUrl: imageUploadUrl, targetPath: targetPath, onUpload: onUpload })));
    }
    // Multi-image mode without drag and drop
    return (React.createElement(React.Fragment, null,
        (currentImages || []).map((image, index) => (React.createElement(Image, { key: image.uuid, image: image, onDelete: onDelete, allowDelete: allowDelete, isFirst: index === 0 }))),
        React.createElement(Upload, { imageUploadUrl: imageUploadUrl, targetPath: targetPath, onUpload: onUpload })));
};
export function ImageUploader({ currentImages = [], isMultiple = true, allowDelete = true, onDelete, onUpload, allowSwap = true, targetPath, onSortEnd }) {
    const [images, setImages] = React.useState(currentImages.map((image) => ({
        uuid: image.uuid,
        url: image.url,
        path: image.path
    })));
    const handleSortEnd = (oldIndex, newIndex) => {
        setImages((items) => {
            return arrayMove(items, oldIndex, newIndex);
        });
        if (onSortEnd) {
            onSortEnd(oldIndex, newIndex);
        }
    };
    const addImage = (imageArray) => {
        if (!isMultiple) {
            // For single image mode, replace the current image
            setImages(imageArray);
        }
        else {
            setImages(images.concat(imageArray));
        }
    };
    const removeImage = (imageUuid) => {
        setImages(images.filter((i) => i.uuid !== imageUuid));
    };
    const onDeleteFn = async (image) => {
        if (onDelete) {
            await onDelete(image);
        }
        removeImage(image.uuid);
    };
    const onUploadFn = async (imageArray) => {
        if (onUpload) {
            await onUpload(imageArray);
        }
        addImage(imageArray);
    };
    const [result] = useQuery({
        query: GetUploadApiQuery
    });
    const { data, fetching, error } = result;
    if (error) {
        return React.createElement("p", { className: "text-critical" },
            "There was an error:",
            error.message);
    }
    else if (fetching) {
        return React.createElement(ImageUploaderSkeleton, { itemCount: isMultiple ? 5 : 1 });
    }
    else {
        return (React.createElement("div", { className: "image-uploader-manager" },
            React.createElement("div", { id: 'image-uploader-wrapper', className: isMultiple ? 'image-list' : '' },
                React.createElement(Images, { currentImages: images, addImage: addImage, imageUploadUrl: data.imageUploadUrl, targetPath: targetPath, onDelete: onDeleteFn, onUpload: onUploadFn, allowDelete: allowDelete, allowSwap: allowSwap && isMultiple, onSortEnd: handleSortEnd, isMultiple: isMultiple }))));
    }
}
