// @evershop/evershop/dist/components/admin/ProductSelector.js
import { SimplePagination } from '@components/common/SimplePagination.js';
import { CheckIcon } from '@heroicons/react/24/outline';
import React from 'react';
import { toast } from 'react-toastify';
import { useQuery } from 'urql';
import { ProductListSkeleton } from './ProductListSkeleton.js';
const SearchQuery = `
  query Query ($filters: [FilterInput!]) {
    products(filters: $filters) {
      items {
        productId
        uuid
        sku
        name
        price {
          regular {
            text
          }
        }
        image {
          url
        }
      }
      total
    }
  }
`;
const isProductSelected = (product, selectedProducts) => {
    return selectedProducts.some((selected) => ((selected === null || selected === void 0 ? void 0 : selected.sku) && selected.sku === product.sku) ||
        ((selected === null || selected === void 0 ? void 0 : selected.uuid) && selected.uuid === product.uuid) ||
        ((selected === null || selected === void 0 ? void 0 : selected.productId) && selected.productId === product.productId));
};
const ProductSelector = ({ onSelect, onUnSelect, selectedProducts }) => {
    var _a, _b;
    const limit = 10;
    const [internalSelectedProducts, setSelectedProducts] = React.useState(selectedProducts || []);
    const [inputValue, setInputValue] = React.useState(null);
    const [loading, setLoading] = React.useState(false);
    const [page, setPage] = React.useState(1);
    const [result, reexecuteQuery] = useQuery({
        query: SearchQuery,
        variables: {
            filters: inputValue
                ? [
                    { key: 'keyword', operation: 'eq', value: inputValue },
                    { key: 'page', operation: 'eq', value: page.toString() },
                    { key: 'limit', operation: 'eq', value: limit.toString() }
                ]
                : [
                    { key: 'limit', operation: 'eq', value: limit.toString() },
                    { key: 'page', operation: 'eq', value: page.toString() }
                ]
        },
        pause: true
    });
    const selectProduct = async (sku, uuid, productId) => {
        setSelectedProducts((prev) => [...prev, { sku, uuid, productId }]);
        try {
            await onSelect(sku, uuid, productId);
        }
        catch (e) {
            toast.error(e.message);
        }
    };
    const unSelectProduct = async (sku, uuid, productId) => {
        if (!onUnSelect) {
            return;
        }
        setSelectedProducts((prev) => prev.filter((product) => (product === null || product === void 0 ? void 0 : product.sku) !== sku));
        try {
            await onUnSelect(sku, uuid, productId);
        }
        catch (e) {
            toast.error(e.message);
        }
    };
    React.useEffect(() => {
        reexecuteQuery({ requestPolicy: 'network-only' });
    }, [page]);
    React.useEffect(() => {
        const timer = setTimeout(() => {
            setLoading(false);
            if (inputValue !== null) {
                reexecuteQuery({ requestPolicy: 'network-only' });
            }
        }, 1500);
        return () => clearTimeout(timer);
    }, [inputValue]);
    const { data, fetching, error } = result;
    if (error) {
        return (React.createElement("p", { className: "text-critical" },
            "There was an error fetching products.",
            error.message));
    }
    return (React.createElement("div", null,
        React.createElement("div", { className: "p-2" },
            React.createElement("div", { className: "form-field" },
                React.createElement("input", { type: "text", value: inputValue || '', placeholder: "Search products", onChange: (e) => {
                        setInputValue(e.target.value);
                        setLoading(true);
                    } }))),
        (fetching || loading) && React.createElement(ProductListSkeleton, null),
        !fetching && data && !loading && (React.createElement("div", { className: "divide-y" },
            data.products.items.length === 0 && (React.createElement("div", { className: "p-2 border border-divider rounded flex justify-center items-center" }, inputValue ? (React.createElement("p", null,
                "No products found for query \"",
                inputValue,
                "\u201D")) : (React.createElement("p", null, "You have no products to display")))),
            data.products.items.map((product) => {
                var _a, _b, _c;
                return (React.createElement("div", { key: product.uuid, className: "grid grid-cols-8 gap-5 py-2 border-divider items-center" },
                    React.createElement("div", { className: "col-span-1" },
                        React.createElement("div", { className: "text-border border border-divider p-2 rounded flex justify-center w-10 h-10" },
                            ((_a = product.image) === null || _a === void 0 ? void 0 : _a.url) && (React.createElement("img", { src: (_b = product.image) === null || _b === void 0 ? void 0 : _b.url, alt: product.name })),
                            !((_c = product.image) === null || _c === void 0 ? void 0 : _c.url) && (React.createElement("svg", { className: "self-center", xmlns: "http://www.w3.org/2000/svg", width: "2rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor" },
                                React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M4 16l4.586-4.586a2 2 0 012.828 0L16 16m-2-2l1.586-1.586a2 2 0 012.828 0L20 14m-6-6h.01M6 20h12a2 2 0 002-2V6a2 2 0 00-2-2H6a2 2 0 00-2 2v12a2 2 0 002 2z" }))))),
                    React.createElement("div", { className: "col-span-5" },
                        React.createElement("h3", null, product.name),
                        React.createElement("p", null, product.sku)),
                    React.createElement("div", { className: "col-span-2 text-right" },
                        !isProductSelected(product, internalSelectedProducts) && (React.createElement("button", { type: "button", className: "button secondary", onClick: async (e) => {
                                e.preventDefault();
                                await selectProduct(product.sku, product.uuid, product.productId);
                            } }, "Select")),
                        isProductSelected(product, internalSelectedProducts) && (React.createElement("a", { className: "button primary", href: "#", onClick: (e) => {
                                e.preventDefault();
                                unSelectProduct(product.sku, product.uuid, product.productId);
                            } },
                            React.createElement(CheckIcon, { width: '1.2rem', height: '1.2rem' }))))));
            }))),
        React.createElement("div", { className: "flex justify-between gap-5 pt-5" },
            React.createElement(SimplePagination, { total: (data === null || data === void 0 ? void 0 : data.products.total) || 0, count: ((_b = (_a = data === null || data === void 0 ? void 0 : data.products) === null || _a === void 0 ? void 0 : _a.items) === null || _b === void 0 ? void 0 : _b.length) || 0, page: page, hasNext: limit * page < (data === null || data === void 0 ? void 0 : data.products.total), setPage: setPage }))));
};
export { ProductSelector };
