// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit+productNew/General.js
import { Card } from '@components/admin/Card.js';
import { CategorySelector } from '@components/admin/CategorySelector.js';
import Area from '@components/common/Area.js';
import { Editor } from '@components/common/form/Editor.js';
import { InputField } from '@components/common/form/InputField.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
import { useQuery } from 'urql';
import './General.scss';
import { useFormContext } from 'react-hook-form';
const SKUPriceWeight = ({ sku, price, weight, setting }) => {
    return (React.createElement("div", { className: "grid grid-cols-3 gap-2 mt-4" },
        React.createElement(InputField, { name: "sku", label: "SKU", placeholder: "Enter SKU", defaultValue: sku, required: true, helperText: _('SKU must be unique') }),
        React.createElement(NumberField, { name: "price", placeholder: "Enter price", label: `Price`, defaultValue: price === null || price === void 0 ? void 0 : price.value, unit: setting.storeCurrency, min: 0, required: true }),
        React.createElement(NumberField, { name: "weight", placeholder: "Enter weight", label: `Weight`, defaultValue: weight === null || weight === void 0 ? void 0 : weight.value, unit: setting.weightUnit, required: true, validation: { min: 1 }, helperText: _('Weight must be a positive number') })));
};
const CategoryQuery = `
  query Query ($id: Int!) {
    category(id: $id) {
      categoryId
      name
      path {
        name
      }
    }
  }
`;
const ProductCategory = ({ categoryId, onChange, onUnassign }) => {
    const { register } = useFormContext();
    const [result] = useQuery({
        query: CategoryQuery,
        variables: {
            id: categoryId
        }
    });
    const { data, fetching, error } = result;
    if (error) {
        return (React.createElement("p", { className: "text-critical" },
            "There was an error fetching categories.",
            error.message));
    }
    if (fetching) {
        return React.createElement("span", null, "Loading...");
    }
    return (React.createElement("div", null,
        data.category.path.map((item, index) => (React.createElement("span", { key: item.name, className: "text-gray-500" },
            item.name,
            index < data.category.path.length - 1 && ' > '))),
        React.createElement("span", { className: "text-interactive pl-5" },
            React.createElement("a", { href: "#", onClick: (e) => {
                    e.preventDefault();
                    onChange();
                } }, "Change"),
            React.createElement("a", { href: "#", onClick: (e) => {
                    e.preventDefault();
                    onUnassign();
                }, className: "text-critical ml-5" }, "Unassign")),
        React.createElement("input", { type: "hidden", ...register('category_id'), value: categoryId })));
};
const CategorySelect = ({ product }) => {
    const [category, setCategory] = React.useState(product ? product.category : null);
    const modal = useModal();
    const onSelect = (categoryId) => {
        setCategory({ categoryId });
        modal.close();
    };
    return (React.createElement("div", { className: "mt-4 relative" },
        React.createElement("div", { className: "mb-2" }, "Category"),
        category && (React.createElement("div", { className: "border rounded border-[#c9cccf] mb-2 p-2" },
            React.createElement(ProductCategory, { categoryId: category.categoryId, onChange: () => modal.open(), onUnassign: () => setCategory(null) }))),
        !category && (React.createElement("a", { href: "#", onClick: (e) => {
                e.preventDefault();
                modal.open();
            }, className: "text-interactive" }, "Select category")),
        React.createElement(Modal, { title: "Select Category", isOpen: modal.isOpen, onClose: modal.close },
            React.createElement(CategorySelector, { onSelect: onSelect, onUnSelect: () => { }, selectedCategories: category ? [category] : [] }))));
};
export default function General({ product, setting, productTaxClasses: { items: taxClasses } }) {
    return (React.createElement(Card, { title: "General" },
        React.createElement(Card.Session, null,
            React.createElement(Area, { id: "productEditGeneral", coreComponents: [
                    {
                        component: {
                            default: (React.createElement(InputField, { name: "name", placeholder: "Enter product name", label: "Product Name", defaultValue: product === null || product === void 0 ? void 0 : product.name, required: true, helperText: _('Product name is required') }))
                        },
                        sortOrder: 10,
                        id: 'name'
                    },
                    {
                        component: {
                            default: (React.createElement(SKUPriceWeight, { sku: (product === null || product === void 0 ? void 0 : product.sku) || '', price: (product === null || product === void 0 ? void 0 : product.price.regular) || {
                                    value: undefined
                                }, weight: (product === null || product === void 0 ? void 0 : product.weight) || { value: undefined }, setting: setting }))
                        },
                        sortOrder: 20,
                        id: 'SKUPriceWeight'
                    },
                    {
                        component: {
                            default: React.createElement(CategorySelect, { product: product })
                        },
                        sortOrder: 22,
                        id: 'category'
                    },
                    {
                        component: {
                            default: (React.createElement(SelectField, { name: "tax_class", label: "Tax Class", options: taxClasses.map((taxClass) => ({
                                    value: taxClass.value,
                                    label: taxClass.text
                                })), defaultValue: (product === null || product === void 0 ? void 0 : product.taxClass) || '', required: true, validation: { required: true } }))
                        },
                        sortOrder: 25,
                        id: 'tax_class'
                    },
                    {
                        component: {
                            default: (React.createElement(Editor, { name: "description", label: "Description", value: product === null || product === void 0 ? void 0 : product.description }))
                        },
                        sortOrder: 30,
                        id: 'description'
                    }
                ] }))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 10
};
export const query = `
  query Query {
    product(id: getContextValue("productId", null)) {
      productId
      uuid
      name
      description
      sku
      taxClass
      price {
        regular {
          value
          currency
        }
      }
      weight {
        value
        unit
      }
      category {
        categoryId
        path {
          name
        }
      }
    }
    setting {
      weightUnit
      storeCurrency
    }
    productTaxClasses: taxClasses {
      items {
        value: taxClassId
        text: name
      }
    }
  }
`;
