// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit+productNew/Media.js
import { Card } from '@components/admin/Card.js';
import { ImageUploader } from '@components/admin/ImageUploader.js';
import React, { useEffect } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
export default function Media({ product }) {
    const control = useFormContext().control;
    const { fields, append, remove, replace } = useFieldArray({
        name: 'images',
        control
    });
    useEffect(() => {
        const images = (product === null || product === void 0 ? void 0 : product.image)
            ? [product.image].concat((product === null || product === void 0 ? void 0 : product.gallery) || [])
            : [];
        replace(images);
    }, []);
    return (React.createElement(Card, { title: "Media" },
        React.createElement(Card.Session, null,
            React.createElement(ImageUploader, { currentImages: (product === null || product === void 0 ? void 0 : product.image) ? [product.image].concat((product === null || product === void 0 ? void 0 : product.gallery) || []) : [], allowDelete: true, allowSwap: true, onDelete: (image) => {
                    const index = fields.findIndex((img) => img.uuid === image.uuid);
                    if (index !== -1) {
                        remove(index);
                    }
                }, onUpload: (images) => {
                    append(images);
                }, onSortEnd: (oldIndex, newIndex) => {
                    const newImages = [...fields];
                    const [movedImage] = newImages.splice(oldIndex, 1);
                    newImages.splice(newIndex, 0, movedImage);
                    replace(newImages);
                }, targetPath: `catalog/${Math.floor(Math.random() * (9999 - 1000)) + 1000}/${Math.floor(Math.random() * (9999 - 1000)) + 1000}` }))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 15
};
export const query = `
  query Query {
    product(id: getContextValue("productId", null)) {
      image {
        uuid
        path
        url
      }
      gallery {
        uuid
        path
        url
      }
    }
  }
`;
