// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit+productNew/Seo.js
import { Card } from '@components/admin/Card.js';
import Area from '@components/common/Area.js';
import { InputField } from '@components/common/form/InputField.js';
import { TextareaField } from '@components/common/form/TextareaField.js';
import React from 'react';
export default function SEO({ product }) {
    const fields = [
        {
            component: {
                default: (React.createElement(InputField, { name: "url_key", label: "URL Key", placeholder: "Enter URL Key", required: true, defaultValue: product === null || product === void 0 ? void 0 : product.urlKey, validation: {
                        required: 'URL Key is required',
                        pattern: {
                            value: /^[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*$/,
                            message: 'URL Key must be lowercase and can only contain letters, numbers, and hyphens'
                        }
                    } }))
            },
            sortOrder: 0
        },
        {
            component: {
                default: (React.createElement(InputField, { name: "meta_title", label: "Meta Title", placeholder: "Enter Meta Title", required: true, defaultValue: product === null || product === void 0 ? void 0 : product.metaTitle, validation: {
                        required: 'Meta Title is required'
                    } }))
            },
            sortOrder: 10
        },
        {
            component: {
                default: (React.createElement(InputField, { type: "hidden", name: "meta_keywords", defaultValue: product === null || product === void 0 ? void 0 : product.metaKeywords }))
            },
            sortOrder: 20
        },
        {
            component: {
                default: (React.createElement(TextareaField, { name: "meta_description", label: "Meta Description", placeholder: "Enter Meta Description", defaultValue: product === null || product === void 0 ? void 0 : product.metaDescription }))
            },
            sortOrder: 30
        }
    ];
    return (React.createElement(Card, { title: "Search engine optimize" },
        React.createElement(Card.Session, null,
            React.createElement(Area, { id: "productEditSeo", coreComponents: fields }))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 60
};
export const query = `
  query Query {
    product(id: getContextValue('productId', null)) {
      urlKey
      metaTitle
      metaKeywords
      metaDescription
    }
  }
`;
