// @evershop/evershop/dist/modules/catalog/pages/admin/categoryEdit/CategoryEditForm.js
import { FormButtons } from '@components/admin/FormButtons.js';
import Area from '@components/common/Area.js';
import { Form } from '@components/common/form/Form.js';
import React from 'react';
export default function CategoryEditForm({ action, gridUrl }) {
    return (React.createElement(Form, { action: action, method: "PATCH", id: "categoryEditForm", submitBtn: false },
        React.createElement("div", { className: "grid grid-cols-3 gap-x-5 grid-flow-row " },
            React.createElement("div", { className: "col-span-2 grid grid-cols-1 gap-5 auto-rows-max" },
                React.createElement(Area, { id: "leftSide", noOuter: true })),
            React.createElement("div", { className: "col-span-1 grid grid-cols-1 gap-5 auto-rows-max" },
                React.createElement(Area, { id: "rightSide", noOuter: true }))),
        React.createElement(FormButtons, { formId: "categoryEditForm", cancelUrl: gridUrl })));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    action: url(routeId: "updateCategory", params: [{key: "id", value: getContextValue("categoryUuid")}]),
    gridUrl: url(routeId: "categoryGrid")
  }
`;
