// @evershop/evershop/dist/modules/catalog/pages/admin/categoryEdit/Products.js
import { Card } from '@components/admin/Card.js';
import { ProductListSkeleton } from '@components/admin/ProductListSkeleton.js';
import { ProductSelector } from '@components/admin/ProductSelector.js';
import Button from '@components/common/Button.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import React from 'react';
import { toast } from 'react-toastify';
import { useQuery } from 'urql';
const ProductsQuery = `
  query Query ($id: Int, $filters: [FilterInput!]) {
    category (id: $id) {
      products (filters: $filters) {
        items {
          productId
          uuid
          name
          sku
          price {
            regular {
              text
            }
          }
          image {
            url
          }
          editUrl
          removeFromCategoryUrl
        }
        total
      }
    }
  }
`;
export default function Products({ category: { categoryId, addProductApi } }) {
    const [keyword, setKeyword] = React.useState('');
    const [loading, setLoading] = React.useState(false);
    const [page, setPage] = React.useState(1);
    const [removing, setRemoving] = React.useState([]);
    const modal = useModal();
    const addProductFunction = async (sku, uuid) => {
        const response = await fetch(addProductApi, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                product_id: uuid
            }),
            credentials: 'include'
        });
        const data = await response.json();
        if (!data.success) {
            toast.error(data.message);
        }
        else {
            reexecuteQuery({ requestPolicy: 'network-only' });
        }
    };
    // Run query again when page changes
    const [result, reexecuteQuery] = useQuery({
        query: ProductsQuery,
        variables: {
            id: categoryId,
            filters: !keyword
                ? [
                    { key: 'page', operation: 'eq', value: page.toString() },
                    { key: 'limit', operation: 'eq', value: '10' }
                ]
                : [
                    { key: 'page', operation: 'eq', value: page.toString() },
                    { key: 'limit', operation: 'eq', value: '10' },
                    { key: 'keyword', operation: 'eq', value: keyword }
                ]
        },
        pause: true
    });
    React.useEffect(() => {
        reexecuteQuery({ requestPolicy: 'network-only' });
    }, []);
    const removeProduct = async (api, uuid) => {
        setRemoving([...removing, uuid]);
        await fetch(api, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json'
            },
            credentials: 'same-origin'
        });
        setPage(1);
        reexecuteQuery({ requestPolicy: 'network-only' });
    };
    React.useEffect(() => {
        const timer = setTimeout(() => {
            reexecuteQuery({ requestPolicy: 'network-only' });
            setLoading(false);
        }, 1500);
        return () => clearTimeout(timer);
    }, [keyword]);
    React.useEffect(() => {
        if (result.fetching) {
            return;
        }
        reexecuteQuery({ requestPolicy: 'network-only' });
    }, [page]);
    const { data, fetching, error } = result;
    return (React.createElement(Card, { title: "Products", actions: [
            {
                name: 'Add products',
                onAction: () => {
                    modal.open();
                }
            }
        ] },
        error && (React.createElement(Card.Session, null,
            React.createElement("span", { className: "text-critical" }, error.message))),
        data && (React.createElement(Modal, { title: "Add Products", onClose: modal.close, isOpen: modal.isOpen },
            React.createElement(ProductSelector, { onSelect: addProductFunction, selectedProducts: data.category.products.items.map((p) => ({
                    sku: p.sku,
                    uuid: p.uuid,
                    productId: p.productId
                })) }))),
        React.createElement(Card.Session, null,
            React.createElement("div", null,
                React.createElement("div", { className: "mb-5" },
                    React.createElement("div", { className: "form-field" },
                        React.createElement("input", { type: "text", value: keyword, placeholder: "Search products", onChange: (e) => {
                                setLoading(true);
                                setKeyword(e.target.value);
                            } }))),
                data && !loading && (React.createElement(React.Fragment, null,
                    data.category.products.items.length === 0 && (React.createElement("div", null, "No product to display.")),
                    React.createElement("div", { className: "flex justify-between" },
                        React.createElement("div", null,
                            React.createElement("i", null,
                                data.category.products.total,
                                " items")),
                        React.createElement("div", null, data.category.products.total > 10 && (React.createElement("div", { className: "flex justify-between gap-2" },
                            page > 1 && (React.createElement("a", { className: "text-interactive", href: "#", onClick: (e) => {
                                    e.preventDefault();
                                    setPage(page - 1);
                                } }, "Previous")),
                            page < data.category.products.total / 10 && (React.createElement("a", { className: "text-interactive", href: "#", onClick: (e) => {
                                    e.preventDefault();
                                    setPage(page + 1);
                                } }, "Next")))))),
                    React.createElement("div", { className: "divide-y" }, data.category.products.items.map((p) => {
                        var _a, _b, _c;
                        return (React.createElement("div", { key: p.uuid, className: "flex justify-between py-2 border-divider items-center" },
                            React.createElement("div", { className: "flex justify-items-start gap-5" },
                                React.createElement("div", { className: "grid-thumbnail text-border border border-divider p-2 rounded flex justify-center w-10 h-10" },
                                    ((_a = p.image) === null || _a === void 0 ? void 0 : _a.url) && (React.createElement("img", { className: "self-center", src: (_b = p.image) === null || _b === void 0 ? void 0 : _b.url, alt: "" })),
                                    !((_c = p.image) === null || _c === void 0 ? void 0 : _c.url) && (React.createElement("svg", { className: "self-center", xmlns: "http://www.w3.org/2000/svg", width: "2rem", fill: "none", viewBox: "0 0 24 24", stroke: "currentColor" },
                                        React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", strokeWidth: 2, d: "M4 16l4.586-4.586a2 2 0 012.828 0L16 16m-2-2l1.586-1.586a2 2 0 012.828 0L20 14m-6-6h.01M6 20h12a2 2 0 002-2V6a2 2 0 00-2-2H6a2 2 0 00-2 2v12a2 2 0 002 2z" })))),
                                React.createElement("div", null,
                                    React.createElement("a", { href: p.editUrl || '', className: "font-semibold hover:underline" }, p.name))),
                            React.createElement("div", { className: "text-right" },
                                React.createElement(Button, { title: "Remove", variant: "danger", onAction: async () => {
                                        await removeProduct(p.removeFromCategoryUrl, p.uuid);
                                    }, isLoading: removing.includes(p.uuid) }))));
                    })))),
                (fetching || loading) && React.createElement(ProductListSkeleton, null)))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 15
};
export const query = `
  query Query {
    category(id: getContextValue("categoryId", null)) {
      categoryId
      addProductApi: addProductUrl
    }
  }
`;
