// @evershop/evershop/dist/components/frontStore/checkout/OrderTotalSummary.js
import Area from '@components/common/Area.js';
import { useAppState } from '@components/common/context/app.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
const Total = ({ total, totalTaxAmount, priceIncludingTax }) => {
    return (React.createElement("div", { className: "summary__row grand-total flex justify-between py-2" },
        (priceIncludingTax && (React.createElement("div", null,
            React.createElement("div", { className: "font-bold" },
                React.createElement("span", null, _('Total'))),
            React.createElement("div", null,
                React.createElement("span", { className: "italic font-normal" },
                    "(",
                    _('Inclusive of tax ${totalTaxAmount}', { totalTaxAmount }),
                    ")"))))) || React.createElement("span", { className: "self-center font-bold" }, _('Total')),
        React.createElement("div", null,
            React.createElement("div", null),
            React.createElement("span", null, total))));
};
const Tax = ({ showPriceIncludingTax, amount }) => {
    if (showPriceIncludingTax) {
        return null;
    }
    return (React.createElement("div", { className: "summary-row flex justify-between py-2" },
        React.createElement("span", null, _('Tax')),
        React.createElement("div", null,
            React.createElement("div", null),
            React.createElement("span", null, amount))));
};
const Subtotal = ({ subTotal }) => {
    return (React.createElement("div", { className: "flex justify-between gap-7 py-2" },
        React.createElement("div", null, _('Sub total')),
        React.createElement("span", null, subTotal)));
};
const Discount = ({ discountAmount, coupon }) => {
    if (!coupon) {
        return null;
    }
    return (React.createElement("div", { className: "flex justify-between gap-7 py-2" },
        React.createElement("div", null, _('Discount(${coupon})', { coupon })),
        React.createElement("span", null,
            "- ",
            discountAmount)));
};
const Shipping = ({ method, cost }) => {
    return (React.createElement("div", { className: "summary-row flex justify-between gap-7 py-2" },
        method && (React.createElement(React.Fragment, null,
            React.createElement("span", null, _('Shipping (${method})', { method })),
            React.createElement("div", null,
                React.createElement("span", { className: "block" }, cost)))),
        !method && (React.createElement(React.Fragment, null,
            React.createElement("span", null, _('Shipping')),
            React.createElement("span", { className: "text-gray-500 italic font-normal" }, _('No shipping is required for this order'))))));
};
const OrderTotalSummary = ({ subTotal, discountAmount, coupon, shippingMethod, shippingCost, taxAmount, total }) => {
    const { config: { tax: { priceIncludingTax } } } = useAppState();
    return (React.createElement("div", { className: "order__total__summary font-semibold" },
        React.createElement(Area, { id: "orderSummaryBeforeSubTotal", noOuter: true }),
        React.createElement(Subtotal, { subTotal: subTotal }),
        React.createElement(Area, { id: "orderSummaryAfterSubTotal", noOuter: true }),
        React.createElement(Area, { id: "orderSummaryBeforeDiscount", noOuter: true }),
        React.createElement(Discount, { discountAmount: discountAmount, coupon: coupon }),
        React.createElement(Area, { id: "orderSummaryAfterDiscount", noOuter: true }),
        React.createElement(Area, { id: "orderSummaryBeforeShipping", noOuter: true }),
        React.createElement(Shipping, { method: shippingMethod, cost: shippingCost }),
        React.createElement(Area, { id: "orderSummaryAfterShipping", noOuter: true }),
        React.createElement(Area, { id: "orderSummaryBeforeTax", noOuter: true }),
        React.createElement(Tax, { amount: taxAmount, showPriceIncludingTax: priceIncludingTax }),
        React.createElement(Area, { id: "orderSummaryAfterTax", noOuter: true }),
        React.createElement(Area, { id: "orderSummaryBeforeTotal", noOuter: true }),
        React.createElement(Total, { total: total, totalTaxAmount: taxAmount, priceIncludingTax: priceIncludingTax }),
        React.createElement(Area, { id: "orderSummaryAfterTotal", noOuter: true })));
};
export { OrderTotalSummary, Subtotal, Discount, Shipping, Tax, Total };
