// @evershop/evershop/dist/modules/checkout/pages/frontStore/checkout/Checkout.js
import Area from '@components/common/Area.js';
import { Form } from '@components/common/form/Form.js';
import { CartItems } from '@components/frontStore/cart/CartItems.js';
import { CartSummaryItemsList } from '@components/frontStore/cart/CartSummaryItems.js';
import { CartTotalSummary } from '@components/frontStore/cart/CartTotalSummary.js';
import { CheckoutButton } from '@components/frontStore/checkout/CheckoutButton.js';
import { CheckoutProvider } from '@components/frontStore/checkout/CheckoutContext.js';
import { ContactInformation } from '@components/frontStore/checkout/ContactInformation.js';
import { Payment } from '@components/frontStore/checkout/Payment.js';
import { Shipment } from '@components/frontStore/checkout/Shipment.js';
import React from 'react';
import './Checkout.scss';
import { useForm } from 'react-hook-form';
export default function CheckoutPage({ placeOrderApi, checkoutSuccessUrl }) {
    const [disabled, setDisabled] = React.useState(false);
    const form = useForm({
        disabled: disabled,
        mode: 'onBlur',
        reValidateMode: 'onBlur',
        defaultValues: {}
    });
    return (React.createElement(CheckoutProvider, { form: form, enableForm: () => setDisabled(false), disableForm: () => setDisabled(true), allowGuestCheckout: true, placeOrderApi: placeOrderApi, checkoutSuccessUrl: checkoutSuccessUrl },
        React.createElement("div", { className: "page-width grid grid-cols-1 md:grid-cols-2 gap-7 pt-8 pb-8" },
            React.createElement(Form, { form: form, submitBtn: false },
                React.createElement("div", null,
                    React.createElement(ContactInformation, null),
                    React.createElement(Shipment, null),
                    React.createElement(Payment, null),
                    React.createElement(CheckoutButton, null)),
                React.createElement(Area, { id: "checkoutForm", noOuter: true })),
            React.createElement("div", null,
                React.createElement(CartItems, null, ({ items, loading, showPriceIncludingTax }) => (React.createElement(CartSummaryItemsList, { items: items, loading: loading, showPriceIncludingTax: showPriceIncludingTax }))),
                React.createElement(CartTotalSummary, null)))));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    placeOrderApi: url(routeId: "createOrder")
    checkoutSuccessUrl: url(routeId: "checkoutSuccess")
  }
`;
