// @evershop/evershop/dist/modules/setting/pages/admin/storeSetting/StoreSetting.js
import { Card } from '@components/admin/Card.js';
import { SettingMenu } from '@components/admin/SettingMenu.js';
import Spinner from '@components/admin/Spinner.js';
import Area from '@components/common/Area.js';
import { EmailField } from '@components/common/form/EmailField.js';
import { Form } from '@components/common/form/Form.js';
import { InputField } from '@components/common/form/InputField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import { TelField } from '@components/common/form/TelField.js';
import { TextareaField } from '@components/common/form/TextareaField.js';
import React from 'react';
import { useQuery } from 'urql';
const ProvincesQuery = `
  query Province($countries: [String]) {
    provinces (countries: $countries) {
      code
      name
      countryCode
    }
  }
`;
const CountriesQuery = `
  query Country($countries: [String]) {
    countries (countries: $countries) {
      code
      name
    }
  }
`;
const Province = ({ selectedCountry = 'US', selectedProvince, allowedCountries = [], fieldName = 'storeProvince' }) => {
    const [result] = useQuery({
        query: ProvincesQuery,
        variables: { countries: allowedCountries }
    });
    const { data, fetching, error } = result;
    if (fetching)
        return (React.createElement("div", null,
            React.createElement(Spinner, null)));
    if (error) {
        return React.createElement("p", { className: "text-red-500" }, error.message);
    }
    const provinces = data.provinces.filter((p) => p.countryCode === selectedCountry);
    if (!provinces.length) {
        return null;
    }
    return (React.createElement("div", null,
        React.createElement(SelectField, { id: "storeProvince", defaultValue: selectedProvince, name: fieldName, label: "Province", placeholder: "Province", required: true, options: provinces.map((p) => ({ value: p.code, label: p.name })) })));
};
const Country = ({ selectedCountry, setSelectedCountry, allowedCountries = [], fieldName = 'storeCountry' }) => {
    const onChange = (e) => {
        setSelectedCountry(e.target.value);
    };
    const [result] = useQuery({
        query: CountriesQuery,
        variables: { countries: allowedCountries }
    });
    const { data, fetching, error } = result;
    if (fetching)
        return (React.createElement("div", null,
            React.createElement(Spinner, null)));
    if (error) {
        return React.createElement("p", { className: "text-red-500" }, error.message);
    }
    return (React.createElement("div", { style: { marginTop: '1rem' } },
        React.createElement(SelectField, { defaultValue: selectedCountry, name: fieldName, label: "Country", placeholder: "Country", onChange: onChange, required: true, options: data.countries.map((c) => ({ value: c.code, label: c.name })) })));
};
const StorePhoneNumber = ({ storePhoneNumber }) => {
    return (React.createElement("div", null,
        React.createElement(TelField, { name: "storePhoneNumber", label: "Store Phone Number", placeholder: "Store Phone Number", defaultValue: storePhoneNumber })));
};
const StoreEmail = ({ storeEmail }) => {
    return (React.createElement("div", null,
        React.createElement(EmailField, { name: "storeEmail", label: "Store Email", placeholder: "Store Email", defaultValue: storeEmail })));
};
export default function StoreSetting({ saveSettingApi, setting: { storeName, storeDescription, storePhoneNumber, storeEmail, storeCountry, storeAddress, storeCity, storeProvince, storePostalCode } }) {
    const [selectedCountry, setSelectedCountry] = React.useState(() => {
        const country = storeCountry;
        if (!country) {
            return 'US';
        }
        else {
            return country;
        }
    });
    return (React.createElement("div", { className: "main-content-inner" },
        React.createElement("div", { className: "grid grid-cols-6 gap-x-5 grid-flow-row " },
            React.createElement("div", { className: "col-span-2" },
                React.createElement(SettingMenu, null)),
            React.createElement("div", { className: "col-span-4" },
                React.createElement(Form, { method: "POST", id: "storeSetting", action: saveSettingApi },
                    React.createElement(Card, null,
                        React.createElement(Card.Session, { title: "Store Information" },
                            React.createElement(Area, { id: "storeInfoSetting", coreComponents: [
                                    {
                                        component: {
                                            default: (React.createElement(InputField, { name: "storeName", label: "Store Name", required: true, placeholder: "Store Name", defaultValue: storeName }))
                                        },
                                        sortOrder: 10
                                    },
                                    {
                                        component: {
                                            default: (React.createElement(TextareaField, { name: "storeDescription", label: "Store Description", placeholder: "Store Description", defaultValue: storeDescription, required: true }))
                                        },
                                        sortOrder: 20
                                    }
                                ], noOuter: true })),
                        React.createElement(Card.Session, { title: "Contact Information" },
                            React.createElement(Area, { id: "storeContactSetting", coreComponents: [
                                    {
                                        component: {
                                            default: StorePhoneNumber
                                        },
                                        props: {
                                            storePhoneNumber
                                        },
                                        sortOrder: 10
                                    },
                                    {
                                        component: {
                                            default: StoreEmail
                                        },
                                        props: {
                                            storeEmail
                                        },
                                        sortOrder: 20
                                    }
                                ], className: "grid grid-cols-2 gap-5 mt-5" })),
                        React.createElement(Card.Session, { title: "Address" },
                            React.createElement(Country, { selectedCountry: storeCountry, setSelectedCountry: setSelectedCountry }),
                            React.createElement(InputField, { name: "storeAddress", label: "Address", defaultValue: storeAddress, placeholder: "Store Address" }),
                            React.createElement("div", { className: "grid grid-cols-3 gap-5 mt-5" },
                                React.createElement("div", null,
                                    React.createElement(InputField, { name: "storeCity", label: "City", defaultValue: storeCity, placeholder: "City" })),
                                React.createElement(Province, { selectedProvince: storeProvince, selectedCountry: selectedCountry }),
                                React.createElement("div", null,
                                    React.createElement(InputField, { name: "storePostalCode", label: "Postal Code", defaultValue: storePostalCode, placeholder: "Postal Code" }))))))))));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    saveSettingApi: url(routeId: "saveSetting")
    setting {
      storeName
      storeDescription
      storeTimeZone
      storePhoneNumber
      storeEmail
      storeCountry
      storeAddress
      storeCity
      storeProvince
      storePostalCode
    }
  }
`;
