// @evershop/evershop/dist/modules/cms/pages/admin/widgetEdit+widgetNew/General.js
import { Card } from '@components/admin/Card.js';
import { InputField } from '@components/common/form/InputField.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { RadioGroupField } from '@components/common/form/RadioGroupField.js';
import React from 'react';
import { useFormContext, Controller } from 'react-hook-form';
import Select from 'react-select';
import CreatableSelect from 'react-select/creatable';
const components = {
    DropdownIndicator: null
};
const createOption = (label) => ({
    label,
    value: label
});
const AreaInput = ({ values, control }) => {
    const [inputValue, setInputValue] = React.useState('');
    const handleKeyDown = (event, onChange, value) => {
        if (!inputValue)
            return;
        switch (event.key) {
            case 'Enter':
            case 'Tab':
                const newOption = createOption(inputValue);
                onChange([...value, newOption]);
                setInputValue('');
                event.preventDefault();
                break;
            default:
                break;
        }
    };
    return (React.createElement(Controller, { name: "area", control: control, defaultValue: values.map((v) => v.value), render: ({ field }) => (React.createElement(CreatableSelect, { components: components, inputValue: inputValue, isClearable: true, isMulti: true, menuIsOpen: false, onChange: (newValue) => {
                const stringArray = newValue
                    ? newValue.map((option) => option.value)
                    : [];
                field.onChange(stringArray);
            }, onInputChange: (newValue) => setInputValue(newValue), onKeyDown: (event) => handleKeyDown(event, (newOptions) => {
                const stringArray = newOptions.map((option) => option.value);
                field.onChange(stringArray);
            }, field.value
                ? field.value.map((val) => typeof val === 'string' ? createOption(val) : val)
                : []), placeholder: "Type area and press enter...", value: field.value
                ? field.value.map((val) => typeof val === 'string' ? createOption(val) : val)
                : [] })) }));
};
export default function General({ widget, routes }) {
    var _a;
    const { register, control } = useFormContext();
    const allRoutes = [
        {
            value: 'all',
            label: 'All',
            isAdmin: false,
            isApi: false,
            method: ['GET']
        },
        ...routes
    ];
    return (React.createElement(Card, null,
        React.createElement(Card.Session, { title: "Name" },
            React.createElement(InputField, { name: "name", defaultValue: widget === null || widget === void 0 ? void 0 : widget.name, required: true, validation: { required: 'Name is required' }, placeholder: "Name" })),
        React.createElement(Card.Session, { title: "Status" },
            React.createElement(RadioGroupField, { name: "status", defaultValue: widget === null || widget === void 0 ? void 0 : widget.status, required: true, validation: { required: 'Status is required' }, options: [
                    { value: 0, label: 'Disabled' },
                    { value: 1, label: 'Enabled' }
                ] })),
        React.createElement(Card.Session, { title: "Area" },
            React.createElement(AreaInput, { control: control, values: ((_a = widget === null || widget === void 0 ? void 0 : widget.area) === null || _a === void 0 ? void 0 : _a.length)
                    ? widget.area.map((a) => ({ value: a, label: a }))
                    : [] })),
        React.createElement(Card.Session, { title: "Page" },
            React.createElement(Controller, { name: "route", control: control, defaultValue: (widget === null || widget === void 0 ? void 0 : widget.route)
                    ? widget.route // Keep as string array
                    : [], render: ({ field }) => (React.createElement(Select, { options: allRoutes.filter((r) => r.isApi === false &&
                        r.isAdmin === false &&
                        r.method.includes('GET') &&
                        r.method.length === 1), hideSelectedOptions: true, isMulti: true, "aria-label": "Select pages", onChange: (selectedOptions) => {
                        const stringArray = selectedOptions
                            ? selectedOptions.map((option) => option.value)
                            : [];
                        field.onChange(stringArray);
                    }, value: allRoutes.filter((r) => { var _a; return (_a = field.value) === null || _a === void 0 ? void 0 : _a.includes(r.value); }), className: "page-select relative z-50" })) })),
        React.createElement(Card.Session, { title: "Sort order" },
            React.createElement(NumberField, { name: "sort_order", defaultValue: widget === null || widget === void 0 ? void 0 : widget.sortOrder, placeholder: "Sort Order", validation: {
                    required: 'Sort order is required',
                    min: {
                        value: 0,
                        message: 'Sort order must be a positive number'
                    }
                }, required: true, helperText: "The order in which this widget will be displayed. Lower numbers appear first." }))));
}
export const layout = {
    areaId: 'rightSide',
    sortOrder: 15
};
export const query = `
  query Query {
    widget(id: getContextValue("widgetId", null)) {
      name
      status
      sortOrder
      area
      route
    }
    routes {
      value: id
      label: name
      isApi
      isAdmin
      method
    }
  }
`;
