// @evershop/evershop/dist/modules/catalog/pages/admin/categoryEdit+categoryNew/General.js
import { Card } from '@components/admin/Card.js';
import { CategoryTree } from '@components/admin/CategoryTree.js';
import Area from '@components/common/Area.js';
import { Editor } from '@components/common/form/Editor.js';
import { InputField } from '@components/common/form/InputField.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import React from 'react';
import './General.scss';
const ParentCategory = ({ parent }) => {
    const [selecting, setSelecting] = React.useState(false);
    const [category, setCategory] = React.useState(parent || null);
    const modal = useModal();
    return (React.createElement("div", { className: "mt-4 relative" },
        React.createElement("div", { className: "mb-2" }, "Parent category"),
        category && (React.createElement("div", { className: "border rounded border-[#c9cccf] mb-2 p-2" },
            category.path.map((item, index) => (React.createElement("span", { key: item.name, className: "text-gray-500" },
                item.name,
                index < category.path.length - 1 && ' > '))),
            React.createElement("span", { className: "text-interactive pl-5 hover:underline" },
                React.createElement("a", { href: "#", onClick: (e) => {
                        e.preventDefault();
                        modal.open();
                    } }, "Change")),
            React.createElement("span", { className: "text-critical pl-5 hover:underline" },
                React.createElement("a", { href: "#", onClick: (e) => {
                        e.preventDefault();
                        setCategory(null);
                    } }, "Unlink")))),
        !selecting && !category && (React.createElement("a", { href: "#", onClick: (e) => {
                e.preventDefault();
                setSelecting(!selecting);
            }, className: "text-interactive" }, "Select category")),
        React.createElement(Modal, { title: "Select a parent category", onClose: modal.close, isOpen: modal.isOpen },
            React.createElement("div", { className: "px-2" },
                React.createElement(CategoryTree, { selectedCategories: category ? [category] : [], onSelect: (c) => {
                        setCategory(c);
                        modal.close();
                    } }))),
        React.createElement(InputField, { type: "hidden", name: "parent_id", value: (category === null || category === void 0 ? void 0 : category.categoryId) || '' })));
};
export default function General({ category }) {
    const fields = [
        {
            component: {
                default: (React.createElement(InputField, { name: "name", label: "Category Name", placeholder: "Enter Category Name", defaultValue: (category === null || category === void 0 ? void 0 : category.name) || '', required: true, validation: {
                        required: 'Category name is required'
                    } }))
            },
            sortOrder: 10,
            id: 'name'
        },
        {
            component: { default: ParentCategory },
            props: {
                parent: category === null || category === void 0 ? void 0 : category.parent,
                currentId: category === null || category === void 0 ? void 0 : category.categoryId
            },
            sortOrder: 15,
            id: 'parent'
        },
        {
            component: {
                default: (React.createElement(Editor, { name: "description", label: "Description", value: (category === null || category === void 0 ? void 0 : category.description) || [] }))
            },
            sortOrder: 30
        }
    ];
    return (React.createElement(Card, { title: "General" },
        React.createElement(Card.Session, null,
            React.createElement(Area, { id: "categoryEditGeneral", coreComponents: fields }))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 10
};
export const query = `
  query Query {
    category(id: getContextValue("categoryId", null)) {
      categoryId
      name
      hasChildren
      description
      status
      parent {
        categoryId
        hasChildren
        name
        path {
          name
        }
      }
    }
  }
`;
