// @evershop/evershop/dist/modules/catalog/pages/admin/categoryEdit+categoryNew/Image.js
import React, { useEffect, useState } from 'react';
import './Image.scss';
import { Card } from '@components/admin/Card.js';
import { ImageUploader } from '@components/admin/ImageUploader.js';
import { InputField } from '@components/common/form/InputField.js';
import { useFormContext } from 'react-hook-form';
export default function Image({ category }) {
    const [image, setImage] = useState(category === null || category === void 0 ? void 0 : category.image);
    const { setValue } = useFormContext();
    useEffect(() => {
        if (image) {
            setValue('image', image.url);
        }
        else {
            setValue('image', '');
        }
    }, [image, setValue]);
    return (React.createElement(Card, { title: "Category banner" },
        React.createElement(Card.Session, null,
            React.createElement(ImageUploader, { onUpload: (images) => {
                    if (images.length > 0) {
                        setImage(images[0]);
                    }
                }, isMultiple: false, allowDelete: true, onDelete: () => {
                    setImage(undefined);
                }, currentImages: image ? [image] : [], targetPath: `catalog/${Math.floor(Math.random() * (9999 - 1000)) + 1000}/${Math.floor(Math.random() * (9999 - 1000)) + 1000}` }),
            React.createElement(InputField, { type: "hidden", defaultValue: image === null || image === void 0 ? void 0 : image.url, name: "image" }))));
}
export const layout = {
    areaId: 'rightSide',
    sortOrder: 10
};
export const query = `
  query Query {
    category(id: getContextValue("categoryId", null)) {
      image {
        id: uuid
        url
      }
    }
  }
`;
