// @evershop/evershop/dist/modules/catalog/pages/admin/categoryEdit+categoryNew/Seo.js
import { Card } from '@components/admin/Card.js';
import Area from '@components/common/Area.js';
import { InputField } from '@components/common/form/InputField.js';
import { TextareaField } from '@components/common/form/TextareaField.js';
import React from 'react';
export default function Seo({ category }) {
    const fields = [
        {
            component: {
                default: (React.createElement(InputField, { name: "url_key", label: "URL key", placeholder: "Enter URL key", defaultValue: (category === null || category === void 0 ? void 0 : category.urlKey) || '', required: true, validation: {
                        required: 'URL key is required',
                        pattern: {
                            value: /^[a-z0-9]+(?:-[a-z0-9]+)*$/,
                            message: 'URL key must be lowercase and can only contain alphanumeric characters and hyphens'
                        }
                    } }))
            },
            sortOrder: 0
        },
        {
            component: {
                default: (React.createElement(InputField, { name: "meta_title", label: "Meta title", placeholder: "Enter Meta title", defaultValue: (category === null || category === void 0 ? void 0 : category.metaTitle) || '', required: true, validation: {
                        required: 'Meta title is required'
                    } }))
            },
            sortOrder: 10
        },
        {
            component: {
                default: (React.createElement(TextareaField, { name: "meta_description", label: "Meta description", placeholder: "Enter Meta description", defaultValue: (category === null || category === void 0 ? void 0 : category.metaDescription) || '', required: true, validation: {
                        required: 'Meta description is required'
                    } }))
            },
            sortOrder: 30
        }
    ];
    return (React.createElement(Card, { title: "Search engine optimize" },
        React.createElement(Card.Session, null,
            React.createElement(Area, { id: "categoryEditSeo", coreComponents: fields }))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 60
};
export const query = `
  query Query {
    category(id: getContextValue('categoryId', null)) {
      urlKey
      metaTitle
      metaKeywords
      metaDescription
    }
  }
`;
