// @evershop/evershop/dist/modules/cms/pages/admin/cmsPageEdit+cmsPageNew/General.js
import { Card } from '@components/admin/Card.js';
import { Editor } from '@components/common/form/Editor.js';
import { InputField } from '@components/common/form/InputField.js';
import { RadioGroupField } from '@components/common/form/RadioGroupField.js';
import React from 'react';
export default function General({ page }) {
    return (React.createElement(Card, { title: "General" },
        React.createElement(Card.Session, null,
            React.createElement("div", { className: "space-y-2" },
                React.createElement("div", null,
                    React.createElement(InputField, { id: "cms_page_name", name: "name", label: "Page Name", placeholder: "Enter page name", defaultValue: page === null || page === void 0 ? void 0 : page.name, required: true, validation: { required: 'Page name is required' }, helperText: "This is the name of the CMS page that will be displayed in the admin panel." })),
                React.createElement("div", { className: "space-y-2" },
                    React.createElement(RadioGroupField, { name: "status", label: "Status", options: [
                            { value: 1, label: 'Enabled' },
                            { value: 0, label: 'Disabled' }
                        ], defaultValue: page === null || page === void 0 ? void 0 : page.status, required: true, helperText: "Enable this page to make it visible on the frontend." })),
                React.createElement("div", null,
                    React.createElement("label", { htmlFor: "content", className: "block mb-2 font-medium" }, "Content"),
                    React.createElement(Editor, { name: "content", value: (page === null || page === void 0 ? void 0 : page.content) || [] }))))));
}
export const layout = {
    areaId: 'wideScreen',
    sortOrder: 10
};
export const query = `
  query Query {
    page: cmsPage(id: getContextValue("cmsPageId", null)) {
      cmsPageId
      name
      status
      sortOrder
      content
    }
  }
`;
