// @evershop/evershop/dist/modules/cms/pages/admin/cmsPageEdit+cmsPageNew/Seo.js
import { Card } from '@components/admin/Card.js';
import { InputField } from '@components/common/form/InputField.js';
import { TextareaField } from '@components/common/form/TextareaField.js';
import React from 'react';
export default function Seo({ page }) {
    return (React.createElement(Card, { title: "Search Engine Optimization" },
        React.createElement(Card.Session, null,
            React.createElement("div", { className: "space-y-2" },
                React.createElement(InputField, { id: "urlKey", name: "url_key", label: "URL Key", placeholder: "Enter URL key", defaultValue: page === null || page === void 0 ? void 0 : page.urlKey, required: true, validation: { required: 'URL key is required' }, helperText: "This is the URL path for the CMS page." }),
                React.createElement(InputField, { id: "metaTitle", name: "meta_title", label: "Meta Title", placeholder: "Enter meta title", defaultValue: page === null || page === void 0 ? void 0 : page.metaTitle, required: true, validation: { required: 'Meta title is required' }, helperText: "This is the meta title for the CMS page." }),
                React.createElement(TextareaField, { name: "meta_description", label: "Meta Description", placeholder: "Enter meta description", defaultValue: page === null || page === void 0 ? void 0 : page.metaDescription })))));
}
export const layout = {
    areaId: 'wideScreen',
    sortOrder: 30
};
export const query = `
  query Query {
    page: cmsPage(id: getContextValue('cmsPageId', null)) {
      urlKey
      metaTitle
      metaKeywords
      metaDescription
    }
  }
`;
