// @evershop/evershop/dist/components/common/form/ReactSelectField.js
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
import { Controller, useFormContext } from 'react-hook-form';
import Select from 'react-select';
export function ReactSelectField({ name, label, error, wrapperClassName = 'form-field', helperText, required, validation, options, className, isMulti = false, defaultValue, ...selectProps }) {
    const { control } = useFormContext();
    const fieldId = `field-${name}`;
    const validationRules = {
        ...validation,
        ...(required &&
            !(validation === null || validation === void 0 ? void 0 : validation.required) && {
            required: _('${field} is required', { field: label || name })
        })
    };
    return (React.createElement("div", { className: `${wrapperClassName} ${className || ''}` },
        label && (React.createElement("label", { htmlFor: fieldId },
            label,
            required && React.createElement("span", { className: "required-indicator" }, "*"))),
        React.createElement(Controller, { name: name, control: control, rules: validationRules, defaultValue: defaultValue, render: ({ field, fieldState }) => {
                var _a;
                const fieldError = error || ((_a = fieldState.error) === null || _a === void 0 ? void 0 : _a.message);
                return (React.createElement("div", { className: fieldError !== undefined ? 'error' : '' },
                    React.createElement(Select, { ...field, ...selectProps, inputId: fieldId, options: options, isMulti: isMulti, value: isMulti
                            ? options.filter((option) => {
                                var _a;
                                return (_a = (field.value || defaultValue || [])) === null || _a === void 0 ? void 0 : _a.includes(option.value);
                            })
                            : options.find((option) => { var _a; return option.value === ((_a = field.value) !== null && _a !== void 0 ? _a : defaultValue); }) || null, onChange: (selectedOption) => {
                            if (isMulti) {
                                const values = selectedOption
                                    ? selectedOption.map((option) => option.value)
                                    : [];
                                field.onChange(values);
                            }
                            else {
                                field.onChange(selectedOption
                                    ? selectedOption.value
                                    : null);
                            }
                        }, classNamePrefix: "react-select", styles: {
                            control: (base, state) => ({
                                ...base,
                                minHeight: 'auto',
                                border: '1px solid #d1d5db',
                                borderRadius: '0.375rem',
                                boxShadow: 'none',
                                transition: 'border-color 0.15s ease-in-out',
                                '&:hover': {
                                    borderColor: '#d1d5db'
                                },
                                ...(state.isFocused && {
                                    borderColor: '#3b82f6',
                                    boxShadow: '0 0 0 1px rgb(59, 130, 246)'
                                })
                            }),
                            input: (base) => ({
                                ...base,
                                '& input': {
                                    boxShadow: 'none !important',
                                    outline: 'none !important'
                                }
                            })
                        } }),
                    fieldError && (React.createElement("p", { id: `${fieldId}-error`, className: "field-error" }, fieldError)),
                    helperText && !fieldError && (React.createElement("p", { id: `${fieldId}-helper`, className: "field-helper" }, helperText))));
            } })));
}
