// @evershop/evershop/dist/components/common/form/ToggleField.js
import { Tooltip } from '@components/common/form/Tooltip.js';
import { getNestedError } from '@components/common/form/utils/getNestedError.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
import { Controller, useFormContext } from 'react-hook-form';
export function ToggleField({ name, label, error, helperText, required, validation, wrapperClassName, disabled = false, defaultValue = false, trueValue = true, falseValue = false, trueLabel = 'Yes', falseLabel = 'No', size = 'md', variant = 'default', onChange }) {
    const { control, formState: { errors } } = useFormContext();
    const fieldError = getNestedError(name, errors, error);
    const fieldId = `field-${name}`;
    const validationRules = {
        ...validation,
        ...(required &&
            !(validation === null || validation === void 0 ? void 0 : validation.required) && {
            required: _('${field} is required', { field: label || name })
        })
    };
    const sizeClasses = {
        sm: {
            toggle: 'h-5 w-9',
            thumb: 'h-4 w-4',
            text: 'text-sm'
        },
        md: {
            toggle: 'h-6 w-11',
            thumb: 'h-5 w-5',
            text: 'text-base'
        },
        lg: {
            toggle: 'h-7 w-14',
            thumb: 'h-6 w-6',
            text: 'text-lg'
        }
    };
    const variantClasses = {
        default: {
            active: 'bg-blue-600',
            inactive: 'bg-gray-200'
        },
        success: {
            active: 'bg-green-600',
            inactive: 'bg-gray-200'
        },
        warning: {
            active: 'bg-yellow-600',
            inactive: 'bg-gray-200'
        },
        danger: {
            active: 'bg-red-600',
            inactive: 'bg-gray-200'
        }
    };
    const sizeClass = sizeClasses[size];
    const variantClass = variantClasses[variant];
    return (React.createElement("div", { className: `form-field ${wrapperClassName}` },
        label && (React.createElement("label", { htmlFor: fieldId, className: "block mb-2 font-medium text-gray-700" },
            label,
            required && React.createElement("span", { className: "required-indicator" }, "*"),
            helperText && React.createElement(Tooltip, { content: helperText, position: "top" }))),
        React.createElement(Controller, { name: name, control: control, rules: validationRules, defaultValue: defaultValue, render: ({ field }) => {
                const isActive = field.value === trueValue;
                return (React.createElement("div", { className: fieldError ? 'error' : '' },
                    React.createElement("div", { className: "flex items-center space-x-3" },
                        React.createElement("button", { type: "button", role: "switch", "aria-checked": isActive, "aria-labelledby": `${fieldId}-label`, disabled: disabled, onClick: () => {
                                field.onChange(isActive ? falseValue : trueValue);
                                onChange === null || onChange === void 0 ? void 0 : onChange(isActive ? falseValue : trueValue);
                            }, className: `
                    relative inline-flex shrink-0 cursor-pointer rounded-full border-2 border-transparent 
                    transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2
                    ${sizeClass.toggle}
                    ${isActive ? variantClass.active : variantClass.inactive}
                    ${disabled ? 'opacity-50 cursor-not-allowed' : ''}
                    ${fieldError ? 'ring-2 ring-red-500' : ''}
                  ` },
                            React.createElement("span", { className: "sr-only" }, label || 'Toggle'),
                            React.createElement("span", { "aria-hidden": "true", className: `
                      pointer-events-none inline-block rounded-full bg-white shadow transform ring-0 
                      transition duration-200 ease-in-out
                      ${sizeClass.thumb}
                      ${isActive ? 'translate-x-5' : 'translate-x-0'}
                    `, style: {
                                    transform: isActive
                                        ? `translateX(${size === 'sm'
                                            ? '1.25rem'
                                            : size === 'md'
                                                ? '1.25rem'
                                                : '1.75rem'})`
                                        : 'translateX(0)'
                                } })),
                        React.createElement("div", { className: `flex items-center space-x-4 ${sizeClass.text}` },
                            React.createElement("span", { className: `font-medium ${isActive ? 'text-gray-900' : 'text-gray-500'}` }, isActive ? trueLabel : falseLabel))),
                    React.createElement("input", { type: "hidden", name: name, value: String(field.value) })));
            } }),
        fieldError && (React.createElement("p", { id: `${fieldId}-error`, className: "field-error" }, fieldError))));
}
