// @evershop/evershop/dist/modules/oms/pages/admin/orderEdit/AddTrackingButton.js
import Button from '@components/common/Button.js';
import { Form } from '@components/common/form/Form.js';
import { InputField } from '@components/common/form/InputField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import React from 'react';
import { useForm } from 'react-hook-form';
export default function AddTrackingButton({ order: { shipment }, carriers }) {
    const modal = useModal();
    const form = useForm();
    if (!shipment) {
        return null;
    }
    else {
        return (React.createElement(React.Fragment, null,
            React.createElement(Button, { title: "Edit Tracking Info", variant: "primary", onAction: () => {
                    modal.open();
                } }),
            React.createElement(Modal, { title: "Edit Tracking Information", onClose: modal.close, isOpen: modal.isOpen },
                React.createElement(Form, { form: form, id: "editTrackingInfo", method: "PATCH", action: shipment.updateShipmentApi, submitBtn: false, onSuccess: () => {
                        location.reload();
                    } },
                    React.createElement("div", { className: "grid grid-cols-2 gap-2" },
                        React.createElement("div", null,
                            React.createElement(InputField, { type: "text", name: "tracking_number", label: "Tracking number", placeholder: "Tracking number", defaultValue: shipment.trackingNumber || '', required: true, validation: {
                                    required: 'Tracking number is required'
                                } })),
                        React.createElement("div", null,
                            React.createElement(SelectField, { name: "carrier", label: "Carrier", defaultValue: shipment.carrier || '', required: true, options: carriers, validation: {
                                    required: 'Carrier is required'
                                } }))),
                    React.createElement("div", { className: "flex justify-end" },
                        React.createElement("div", { className: "grid grid-cols-2 gap-2" },
                            React.createElement(Button, { title: "Cancel", variant: "danger", onAction: modal.close }),
                            React.createElement(Button, { title: "Save", variant: "primary", isLoading: form.formState.isSubmitting, onAction: async () => {
                                    document.getElementById('editTrackingInfo').dispatchEvent(new Event('submit', { cancelable: true, bubbles: true }));
                                } })))))));
    }
}
export const layout = {
    areaId: 'order_actions',
    sortOrder: 5
};
export const query = `
  query Query {
    order(uuid: getContextValue("orderId")) {
      shipment {
        shipmentId
        carrier
        trackingNumber
        updateShipmentApi
      }
      createShipmentApi
    },
    carriers {
      label: name
      value: code
    }
  }
`;
