// @evershop/evershop/dist/modules/oms/pages/admin/orderEdit/CancelButton.js
import Button from '@components/common/Button.js';
import { Form } from '@components/common/form/Form.js';
import { TextareaField } from '@components/common/form/TextareaField.js';
import { Modal } from '@components/common/modal/Modal.js';
import { useModal } from '@components/common/modal/useModal.js';
import RenderIfTrue from '@components/common/RenderIfTrue.js';
import React from 'react';
import { useForm } from 'react-hook-form';
import { toast } from 'react-toastify';
export default function CancelButton({ order: { cancelApi, paymentStatus, shipmentStatus } }) {
    const modal = useModal();
    const form = useForm();
    return (React.createElement(RenderIfTrue, { condition: paymentStatus.isCancelable !== false &&
            shipmentStatus.isCancelable !== false },
        React.createElement(Button, { title: "Cancel Order", variant: "danger", onAction: () => {
                modal.open();
            } }),
        React.createElement(Modal, { title: "Cancel Order", onClose: modal.close, isOpen: modal.isOpen },
            React.createElement(Form, { form: form, id: "cancelReason", method: "POST", action: cancelApi, submitBtn: false, onSuccess: (response) => {
                    if (response.error) {
                        toast.error(response.error.message);
                    }
                    else {
                        // Reload the page
                        window.location.reload();
                    }
                } },
                React.createElement("div", null,
                    React.createElement(TextareaField, { name: "reason", label: "Reason for cancellation", placeholder: "Reason for cancellation", required: true, validation: {
                            required: 'Reason is required'
                        } })),
                React.createElement("div", { className: "flex justify-end" },
                    React.createElement("div", { className: "grid grid-cols-2 gap-2" },
                        React.createElement(Button, { title: "Cancel", variant: "danger", onAction: modal.close }),
                        React.createElement(Button, { title: "Submit Cancellation", variant: "primary", isLoading: form.formState.isSubmitting, onAction: async () => {
                                document.getElementById('cancelReason').dispatchEvent(new Event('submit', { cancelable: true, bubbles: true }));
                            } })))))));
}
export const layout = {
    areaId: 'pageHeadingRight',
    sortOrder: 35
};
export const query = `
  query Query {
    order(uuid: getContextValue("orderId")) {
      paymentStatus {
        code
        isCancelable
      }
      shipmentStatus {
        code
        isCancelable
      }
      cancelApi
    }
  }
`;
