// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit/VariantGroup.js
import { Card } from '@components/admin/Card.js';
import React from 'react';
import { New } from './variants/New.js';
import { Variants } from './variants/Variants.js';
const VariantGroup = ({ product, createVariantGroupApi, createProductApi }) => {
    const [group, setGroup] = React.useState((product === null || product === void 0 ? void 0 : product.variantGroup) || null);
    return (React.createElement(Card, { title: "Variant" },
        !group && (React.createElement(New, { currentProductUuid: product.uuid, createVariantGroupApi: createVariantGroupApi, setGroup: setGroup })),
        group && (React.createElement(Variants, { productId: product.productId, productUuid: product.uuid, variantGroup: group, createProductApi: createProductApi }))));
};
export const layout = {
    areaId: 'leftSide',
    sortOrder: 70
};
export const query = `
query Query {
  product(id: getContextValue('productId', null)) {
    productId
    uuid
    variantGroup {
      variantGroupId
      attributes: variantAttributes {
        attributeId
        attributeCode
        attributeName
        options {
          optionId
          optionText
        }
      }
      addItemApi
    }
  }
  createVariantGroupApi: url(routeId: "createVariantGroup")
  createProductApi: url(routeId: "createProduct")
}
`;
export default VariantGroup;
