// @evershop/evershop/dist/modules/catalog/pages/admin/collectionEdit+collectionNew/General.js
import { Card } from '@components/admin/Card.js';
import Area from '@components/common/Area.js';
import { Editor } from '@components/common/form/Editor.js';
import { InputField } from '@components/common/form/InputField.js';
import './General.scss';
import React from 'react';
export default function General({ collection }) {
    const fields = [
        {
            component: {
                default: (React.createElement(InputField, { name: "name", label: "Collection Name", placeholder: "Enter Collection Name", defaultValue: (collection === null || collection === void 0 ? void 0 : collection.name) || '', required: true }))
            },
            sortOrder: 10,
            id: 'name'
        },
        {
            component: {
                default: (React.createElement(InputField, { name: "code", label: "Collection Code", defaultValue: (collection === null || collection === void 0 ? void 0 : collection.code) || '', validation: {
                        required: 'Collection code is required',
                        pattern: {
                            value: /^[a-zA-Z0-9_-]+$/,
                            message: 'Collection code must be alphanumeric and can include underscores or dashes.'
                        }
                    }, placeholder: "Collection Code" }))
            },
            sortOrder: 15,
            id: 'code'
        },
        {
            component: {
                default: (React.createElement(Editor, { name: "description", label: "Description", value: (collection === null || collection === void 0 ? void 0 : collection.description) || [] }))
            },
            sortOrder: 30
        }
    ];
    return (React.createElement(Card, { title: "General" },
        React.createElement(Card.Session, null,
            React.createElement(Area, { id: "collectionEditGeneral", coreComponents: fields }))));
}
export const layout = {
    areaId: 'collectionFormInner',
    sortOrder: 10
};
export const query = `
  query Query {
    collection(code: getContextValue("collectionCode", null)) {
      collectionId
      name
      code
      description
    }
  }
`;
