// @evershop/evershop/dist/modules/checkout/pages/frontStore/cart/ShoppingCart.js
import Area from '@components/common/Area.js';
import Button from '@components/common/Button.js';
import { useCartState } from '@components/frontStore/cart/CartContext.js';
import { CartItems } from '@components/frontStore/cart/CartItems.js';
import { CartTotalSummary } from '@components/frontStore/cart/CartTotalSummary.js';
import { DefaultCartItemList } from '@components/frontStore/cart/DefaultCartItemList.js';
import { ShoppingCartEmpty } from '@components/frontStore/cart/ShoppingCartEmpty.js';
import { _ } from '@evershop/evershop/lib/locale/translate/_';
import React from 'react';
const Title = ({ title }) => {
    return (React.createElement("div", { className: "mb-7 text-center shopping-cart-header" },
        React.createElement("h1", { className: "shopping-cart-title mb-2" }, title),
        React.createElement("a", { href: "/", className: "underline" }, _('Continue Shopping'))));
};
export default function ShoppingCart({ checkoutUrl }) {
    const { data: cart } = useCartState();
    return (React.createElement("div", { className: "cart page-width" }, cart.items.length > 0 ? (React.createElement(React.Fragment, null,
        React.createElement(Title, { title: _('Shopping Cart') }),
        React.createElement("div", { className: "grid gap-10 grid-cols-1 md:grid-cols-4" },
            React.createElement("div", { className: "col-span-1 md:col-span-3" },
                React.createElement(Area, { id: "shoppingCartBeforeItems", noOuter: true }),
                React.createElement(CartItems, null, ({ items, showPriceIncludingTax, loading, onRemoveItem }) => (React.createElement(DefaultCartItemList, { items: items, showPriceIncludingTax: showPriceIncludingTax, loading: loading, onRemoveItem: onRemoveItem }))),
                React.createElement(Area, { id: "shoppingCartAfterItems", noOuter: true })),
            React.createElement("div", { className: "col-span-1 md:col-span-1" },
                React.createElement(Area, { id: "shoppingCartBeforeSummary", noOuter: true }),
                React.createElement("div", { className: "grid grid-cols-1 gap-5 cart-summary" },
                    React.createElement("h4", null, _('Order summary')),
                    React.createElement(CartTotalSummary, null)),
                React.createElement(Area, { id: "shoppingCartBeforeCheckoutButton", noOuter: true }),
                React.createElement("div", { className: "shopping-cart-checkout-btn flex justify-between mt-5" },
                    React.createElement(Button, { url: checkoutUrl, title: _('CHECKOUT'), variant: "primary" })),
                React.createElement(Area, { id: "shoppingCartAfterSummary", noOuter: true }))))) : (React.createElement(ShoppingCartEmpty, null))));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    checkoutUrl: url(routeId: "checkout")
  }
`;
