// @evershop/evershop/dist/modules/cms/pages/admin/widgetNew/WidgetNewForm.js
import { FormButtons } from '@components/admin/FormButtons.js';
import Area from '@components/common/Area.js';
import { Form } from '@components/common/form/Form.js';
import { InputField } from '@components/common/form/InputField.js';
import React from 'react';
import { toast } from 'react-toastify';
export default function WidgetNewForm({ action, gridUrl, type }) {
    return (React.createElement(Form, { action: action, method: "POST", onSuccess: (response) => {
            toast.success('Widget created successfully!');
            setTimeout(() => {
                const editUrl = response.data.links.find((link) => link.rel === 'edit').href;
                window.location.href = editUrl;
            }, 1500);
        }, id: "widgetNewForm", submitBtn: false },
        React.createElement(InputField, { type: "hidden", name: "type", defaultValue: type.code }),
        React.createElement("div", { className: "grid grid-cols-3 gap-x-5 grid-flow-row " },
            React.createElement("div", { className: "col-span-2 grid grid-cols-1 gap-5 auto-rows-max" },
                React.createElement(Area, { id: "leftSide", type: type, noOuter: true })),
            React.createElement("div", { className: "col-span-1 grid grid-cols-1 gap-5 auto-rows-max" },
                React.createElement(Area, { id: "rightSide", type: type, noOuter: true }))),
        React.createElement(FormButtons, { formId: "widgetNewForm", cancelUrl: gridUrl })));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    action: url(routeId: "createWidget")
    gridUrl: url(routeId: "widgetGrid")
    type: widgetType(code: getContextValue('type')) {
      code
      description
      settingComponent
      defaultSetting
    }
  }
`;
