// @evershop/evershop/dist/modules/cms/components/SlideshowSetting.js
/* eslint-disable jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
import { FileBrowser } from '@components/admin/FileBrowser.js';
import { InputField } from '@components/common/form/InputField.js';
import React, { useEffect } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
import { v4 as uuidv4 } from 'uuid';
export default function SlideshowSetting({ slideshowWidget }) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
    const { slides = [], autoplay = true, autoplaySpeed = 3000, arrows = true, dots = true, fullWidth = true, widthValue = 1920, heightValue = 800, heightType = 'auto' } = slideshowWidget || {};
    const { control, setValue, watch } = useFormContext();
    const { fields, append, remove, move } = useFieldArray({
        control,
        name: 'settings.slides'
    });
    const currentSlides = watch('settings.slides', slides);
    const currentAutoplay = watch('settings.autoplay', autoplay);
    const currentAutoplaySpeed = watch('settings.autoplaySpeed', autoplaySpeed);
    const currentArrows = watch('settings.arrows', arrows);
    const currentDots = watch('settings.dots', dots);
    const currentFullWidth = watch('settings.fullWidth', fullWidth);
    useEffect(() => {
        // Initialize slides with existing data
        setValue('settings.slides', (currentSlides === null || currentSlides === void 0 ? void 0 : currentSlides.length) ? currentSlides : []);
        // Initialize the autoplay settings
        const handleAutoplay = currentAutoplay === undefined || currentAutoplay === null
            ? autoplay
            : Boolean(currentAutoplay);
        setValue('settings.autoplay', handleAutoplay);
        // Initialize the autoplay speed
        const speed = Number(currentAutoplaySpeed) || Number(autoplaySpeed) || 3000;
        setValue('settings.autoplaySpeed', speed);
        // Initialize the arrows setting
        const handleArrows = currentArrows === undefined || currentArrows === null
            ? arrows
            : Boolean(currentArrows);
        setValue('settings.arrows', handleArrows);
        // Initialize the dots setting
        const handleDots = currentDots === undefined || currentDots === null
            ? dots
            : Boolean(currentDots);
        setValue('settings.dots', handleDots);
        // Initialize the fullWidth setting
        const handleFullWidth = currentFullWidth === undefined || currentFullWidth === null
            ? fullWidth
            : Boolean(currentFullWidth);
        setValue('settings.fullWidth', handleFullWidth);
        // Always use adaptive height for the slideshow
        setValue('settings.heightType', 'auto');
        // Process all slides to detect image dimensions if they don't have them yet
        if (currentSlides === null || currentSlides === void 0 ? void 0 : currentSlides.length) {
            currentSlides.forEach((slide, index) => {
                if (slide.image && (!slide.width || !slide.height)) {
                    getImageDimensions(slide.image, index);
                }
            });
        }
    }, []);
    const [activeSlideIndex, setActiveSlideIndex] = React.useState(null);
    const [openFileBrowser, setOpenFileBrowser] = React.useState(false);
    // Function to get image dimensions
    const getImageDimensions = (imageUrl, slideIndex) => {
        if (!imageUrl)
            return;
        const img = new Image();
        img.onload = () => {
            const width = img.naturalWidth;
            const height = img.naturalHeight;
            // Update the current slides with the new dimensions
            const newSlides = [...currentSlides];
            newSlides[slideIndex] = {
                ...newSlides[slideIndex],
                width,
                height
            };
            setValue('settings.slides', newSlides);
        };
        img.src = imageUrl;
    };
    const handleImageSelect = (image) => {
        if (activeSlideIndex !== null) {
            setValue(`settings.slides.${activeSlideIndex}.image`, image);
            // Detect image dimensions when a new image is selected
            getImageDimensions(image, activeSlideIndex);
            setOpenFileBrowser(false);
        }
    };
    const addSlide = () => {
        const newSlide = {
            id: uuidv4(),
            image: '',
            width: 0, // Will be automatically set when image is selected
            height: 0, // Will be automatically set when image is selected
            headline: '',
            subText: '',
            buttonText: '',
            buttonLink: '',
            buttonColor: '#3B82F6' // Default blue color
        };
        append(newSlide);
        setTimeout(() => {
            setActiveSlideIndex(fields.length);
        }, 50);
    };
    const moveUp = (index) => {
        if (index > 0) {
            move(index, index - 1);
            setActiveSlideIndex(index - 1);
        }
    };
    const moveDown = (index) => {
        if (index < fields.length - 1) {
            move(index, index + 1);
            setActiveSlideIndex(index + 1);
        }
    };
    return (React.createElement("div", { className: "slideshow-widget" },
        openFileBrowser && (React.createElement("div", { className: "max-h-96" },
            React.createElement(FileBrowser, { isMultiple: false, onInsert: handleImageSelect, close: () => setOpenFileBrowser(false) }))),
        React.createElement("div", { className: "bg-white p-4 rounded shadow-sm mb-4" },
            React.createElement("h2", { className: "text-lg font-medium mb-4" }, "Slideshow Settings"),
            React.createElement("div", { className: "grid grid-cols-2 gap-4" },
                React.createElement("div", { className: "col-span-2 md:col-span-1" },
                    React.createElement("div", { className: "flex items-center mb-4" },
                        React.createElement("input", { type: "checkbox", id: "autoplay", checked: Boolean(currentAutoplay), onChange: (e) => {
                                const isChecked = Boolean(e.target.checked);
                                setValue('settings.autoplay', isChecked);
                            }, className: "mr-2 h-4 w-4" }),
                        React.createElement("label", { htmlFor: "autoplay", className: "text-sm" }, "Autoplay Slides")),
                    Boolean(currentAutoplay) && (React.createElement(InputField, { type: "number", label: "Autoplay Speed (ms)", name: "settings.autoplaySpeed", defaultValue: Number(autoplaySpeed) || 3000, placeholder: "e.g., 3000 for 3 seconds", validation: {
                            min: { value: 1000, message: 'Minimum speed is 1000ms' }
                        } }))),
                React.createElement("div", { className: "col-span-2 md:col-span-1" },
                    React.createElement("div", { className: "flex items-center mb-4" },
                        React.createElement("input", { type: "checkbox", id: "arrows", checked: Boolean(currentArrows), onChange: (e) => {
                                const isChecked = Boolean(e.target.checked);
                                setValue('settings.arrows', isChecked);
                            }, className: "mr-2 h-4 w-4" }),
                        React.createElement("label", { htmlFor: "arrows", className: "text-sm" }, "Show Navigation Arrows"))))),
        React.createElement("div", { className: "mb-4" },
            React.createElement("div", { className: "flex justify-between items-center mb-2" },
                React.createElement("h2", { className: "text-lg font-medium" }, "Slides"),
                React.createElement("button", { type: "button", onClick: addSlide, className: "bg-blue-500 hover:bg-blue-600 text-white px-4 py-2 rounded" }, "Add New Slide")),
            fields.length > 0 ? (React.createElement("div", { className: "grid grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4 mb-4" }, fields.map((slide, index) => {
                var _a, _b;
                return (React.createElement("div", { key: slide.id, onClick: () => setActiveSlideIndex(index), className: `relative border rounded overflow-hidden cursor-pointer ${activeSlideIndex === index ? 'ring-2 ring-blue-500' : ''}` },
                    React.createElement("div", { className: "aspect-[16/9] bg-gray-100 flex items-center justify-center" }, ((_a = currentSlides[index]) === null || _a === void 0 ? void 0 : _a.image) ? (React.createElement("img", { src: currentSlides[index].image, alt: `Slide ${index + 1}`, className: "w-full h-full object-cover" })) : (React.createElement("div", { className: "text-gray-400" }, "No Image"))),
                    React.createElement("div", { className: "p-2 bg-white border-t" },
                        React.createElement("p", { className: "text-sm font-medium truncate" }, ((_b = currentSlides[index]) === null || _b === void 0 ? void 0 : _b.headline) || `Slide ${index + 1}`),
                        React.createElement("div", { className: "flex mt-2" },
                            React.createElement("button", { type: "button", onClick: (e) => {
                                    e.stopPropagation();
                                    moveUp(index);
                                }, disabled: index === 0, className: `mr-1 p-1 rounded ${index === 0
                                    ? 'text-gray-300'
                                    : 'text-gray-500 hover:bg-gray-100'}` },
                                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" },
                                    React.createElement("path", { d: "M18 15l-6-6-6 6" }))),
                            React.createElement("button", { type: "button", onClick: (e) => {
                                    e.stopPropagation();
                                    moveDown(index);
                                }, disabled: index === fields.length - 1, className: `mr-1 p-1 rounded ${index === fields.length - 1
                                    ? 'text-gray-300'
                                    : 'text-gray-500 hover:bg-gray-100'}` },
                                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" },
                                    React.createElement("path", { d: "M6 9l6 6 6-6" }))),
                            React.createElement("button", { type: "button", onClick: (e) => {
                                    e.stopPropagation();
                                    remove(index);
                                    if (activeSlideIndex === index) {
                                        setActiveSlideIndex(null);
                                    }
                                    else if (activeSlideIndex !== null &&
                                        activeSlideIndex > index) {
                                        setActiveSlideIndex(activeSlideIndex - 1);
                                    }
                                }, className: "ml-auto p-1 text-red-500 hover:bg-red-50 rounded" },
                                React.createElement("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" },
                                    React.createElement("path", { d: "M3 6h18" }),
                                    React.createElement("path", { d: "M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6" }),
                                    React.createElement("path", { d: "M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2" })))))));
            }))) : (React.createElement("div", { className: "bg-gray-50 border border-dashed border-gray-300 rounded-lg p-8 text-center mb-4" },
                React.createElement("p", { className: "text-gray-500 mb-4" }, "No slides have been added yet."),
                React.createElement("button", { type: "button", onClick: addSlide, className: "bg-blue-500 hover:bg-blue-600 text-white px-4 py-2 rounded" }, "Add Your First Slide")))),
        activeSlideIndex !== null && fields[activeSlideIndex] && (React.createElement("div", { className: "bg-white p-4 rounded shadow-sm" },
            React.createElement("h3", { className: "text-lg font-medium mb-4" },
                "Edit Slide ",
                activeSlideIndex + 1),
            React.createElement("div", { className: "mb-4 border rounded overflow-hidden" },
                React.createElement("div", { className: "aspect-[16/9] bg-gray-100 relative" },
                    ((_a = currentSlides[activeSlideIndex]) === null || _a === void 0 ? void 0 : _a.image) ? (React.createElement("div", { className: "relative w-full h-full" },
                        React.createElement("img", { src: currentSlides[activeSlideIndex].image, alt: `Slide ${activeSlideIndex + 1}`, className: "w-full h-full object-cover", onLoad: (e) => {
                                var _a, _b;
                                // Additional dimensions detection when the preview image loads
                                const img = e.target;
                                if (img.naturalWidth > 0 && img.naturalHeight > 0) {
                                    if (!((_a = currentSlides[activeSlideIndex]) === null || _a === void 0 ? void 0 : _a.width) ||
                                        !((_b = currentSlides[activeSlideIndex]) === null || _b === void 0 ? void 0 : _b.height)) {
                                        const newSlides = [...currentSlides];
                                        newSlides[activeSlideIndex] = {
                                            ...newSlides[activeSlideIndex],
                                            width: img.naturalWidth,
                                            height: img.naturalHeight
                                        };
                                        setValue('settings.slides', newSlides);
                                    }
                                }
                            } }),
                        React.createElement("div", { className: "absolute inset-0 bg-black bg-opacity-20 flex flex-col items-center justify-center p-4 text-center" },
                            ((_b = currentSlides[activeSlideIndex]) === null || _b === void 0 ? void 0 : _b.headline) && (React.createElement("h3", { className: "text-white text-xl md:text-2xl font-bold mb-2" }, currentSlides[activeSlideIndex].headline)),
                            ((_c = currentSlides[activeSlideIndex]) === null || _c === void 0 ? void 0 : _c.subText) && (React.createElement("p", { className: "text-white mb-4" }, currentSlides[activeSlideIndex].subText)),
                            ((_d = currentSlides[activeSlideIndex]) === null || _d === void 0 ? void 0 : _d.buttonText) && (React.createElement("button", { type: "button", className: "px-4 py-2 rounded", style: {
                                    backgroundColor: currentSlides[activeSlideIndex].buttonColor ||
                                        '#3B82F6'
                                } }, currentSlides[activeSlideIndex].buttonText))))) : (React.createElement("div", { className: "w-full h-full flex items-center justify-center" },
                        React.createElement("button", { type: "button", onClick: () => setOpenFileBrowser(true), className: "bg-blue-500 hover:bg-blue-600 text-white px-4 py-2 rounded" }, "Select Image"))),
                    ((_e = currentSlides[activeSlideIndex]) === null || _e === void 0 ? void 0 : _e.image) && (React.createElement("button", { type: "button", onClick: () => setOpenFileBrowser(true), className: "absolute bottom-2 right-2 bg-white p-2 rounded shadow hover:bg-gray-100" }, "Change Image")))),
            React.createElement("div", { className: "grid grid-cols-1 md:grid-cols-2 gap-4" },
                React.createElement("input", { type: "hidden", name: `settings.slides.${activeSlideIndex}.image`, value: (currentSlides && ((_f = currentSlides[activeSlideIndex]) === null || _f === void 0 ? void 0 : _f.image)) || '' }),
                React.createElement("input", { type: "hidden", name: `settings.slides.${activeSlideIndex}.id`, value: (currentSlides && ((_g = currentSlides[activeSlideIndex]) === null || _g === void 0 ? void 0 : _g.id)) ||
                        uuidv4() }),
                React.createElement("input", { type: "hidden", name: `settings.slides.${activeSlideIndex}.width`, value: ((_h = currentSlides[activeSlideIndex]) === null || _h === void 0 ? void 0 : _h.width) || 0 }),
                React.createElement("input", { type: "hidden", name: `settings.slides.${activeSlideIndex}.height`, value: ((_j = currentSlides[activeSlideIndex]) === null || _j === void 0 ? void 0 : _j.height) || 0 }),
                ((_k = currentSlides[activeSlideIndex]) === null || _k === void 0 ? void 0 : _k.image) && (React.createElement("div", { className: "md:col-span-2 mb-2" },
                    React.createElement("div", { className: "text-sm text-gray-500" }, ((_l = currentSlides[activeSlideIndex]) === null || _l === void 0 ? void 0 : _l.width) &&
                        ((_m = currentSlides[activeSlideIndex]) === null || _m === void 0 ? void 0 : _m.height) ? (React.createElement("p", null,
                        "Image dimensions: ",
                        currentSlides[activeSlideIndex].width,
                        ' ',
                        "\u00D7 ",
                        currentSlides[activeSlideIndex].height,
                        " pixels")) : (React.createElement("p", null, "Detecting image dimensions..."))))),
                React.createElement("div", { className: "md:col-span-2" },
                    React.createElement("label", { className: "block mb-1 text-sm" }, "Headline"),
                    React.createElement("input", { type: "text", className: "w-full p-2 border border-gray-300 rounded", name: `settings.slides.${activeSlideIndex}.headline`, value: ((_o = currentSlides[activeSlideIndex]) === null || _o === void 0 ? void 0 : _o.headline) || '', onChange: (e) => {
                            const newSlides = [...currentSlides];
                            newSlides[activeSlideIndex] = {
                                ...newSlides[activeSlideIndex],
                                headline: e.target.value
                            };
                            setValue('settings.slides', newSlides);
                        }, placeholder: "e.g., New Collection Available" })),
                React.createElement("div", { className: "md:col-span-2" },
                    React.createElement("label", { className: "block mb-1 text-sm" }, "Sub Text"),
                    React.createElement("textarea", { className: "w-full p-2 border border-gray-300 rounded", name: `settings.slides.${activeSlideIndex}.subText`, value: ((_p = currentSlides[activeSlideIndex]) === null || _p === void 0 ? void 0 : _p.subText) || '', onChange: (e) => {
                            const newSlides = [...currentSlides];
                            newSlides[activeSlideIndex] = {
                                ...newSlides[activeSlideIndex],
                                subText: e.target.value
                            };
                            setValue('settings.slides', newSlides);
                        }, placeholder: "e.g., Check out our latest products with special discounts", rows: 3 })),
                React.createElement("div", null,
                    React.createElement("label", { className: "block mb-1 text-sm" }, "Button Text"),
                    React.createElement("input", { type: "text", className: "w-full p-2 border border-gray-300 rounded", name: `settings.slides.${activeSlideIndex}.buttonText`, value: ((_q = currentSlides[activeSlideIndex]) === null || _q === void 0 ? void 0 : _q.buttonText) || '', onChange: (e) => {
                            const newSlides = [...currentSlides];
                            newSlides[activeSlideIndex] = {
                                ...newSlides[activeSlideIndex],
                                buttonText: e.target.value
                            };
                            setValue('settings.slides', newSlides);
                        }, placeholder: "e.g., Shop Now" })),
                React.createElement("div", null,
                    React.createElement("label", { className: "block mb-1 text-sm" }, "Button Link"),
                    React.createElement("input", { type: "text", className: "w-full p-2 border border-gray-300 rounded", name: `settings.slides.${activeSlideIndex}.buttonLink`, value: ((_r = currentSlides[activeSlideIndex]) === null || _r === void 0 ? void 0 : _r.buttonLink) || '', onChange: (e) => {
                            const newSlides = [...currentSlides];
                            newSlides[activeSlideIndex] = {
                                ...newSlides[activeSlideIndex],
                                buttonLink: e.target.value
                            };
                            setValue('settings.slides', newSlides);
                        }, placeholder: "e.g., /category/new-arrivals" })),
                React.createElement("div", null,
                    React.createElement("label", { className: "block mb-1 text-sm" }, "Button Color"),
                    React.createElement("div", { className: "flex items-center" },
                        React.createElement("input", { type: "color", value: ((_s = currentSlides[activeSlideIndex]) === null || _s === void 0 ? void 0 : _s.buttonColor) || '#3B82F6', onChange: (e) => {
                                const newSlides = [...currentSlides];
                                newSlides[activeSlideIndex] = {
                                    ...newSlides[activeSlideIndex],
                                    buttonColor: e.target.value
                                };
                                setValue('settings.slides', newSlides);
                            }, className: "w-10 h-10 rounded mr-2 cursor-pointer" }),
                        React.createElement("input", { type: "text", className: "w-full p-2 border border-gray-300 rounded", value: ((_t = currentSlides[activeSlideIndex]) === null || _t === void 0 ? void 0 : _t.buttonColor) || '#3B82F6', onChange: (e) => {
                                const newSlides = [...currentSlides];
                                newSlides[activeSlideIndex] = {
                                    ...newSlides[activeSlideIndex],
                                    buttonColor: e.target.value
                                };
                                setValue('settings.slides', newSlides);
                            }, placeholder: "#3B82F6" }))))))));
}
export const query = `
  query Query($slides: [SlideInput], $autoplay: Boolean, $autoplaySpeed: Int, $arrows: Boolean, $dots: Boolean) {
    slideshowWidget(
      slides: $slides, 
      autoplay: $autoplay, 
      autoplaySpeed: $autoplaySpeed, 
      arrows: $arrows, 
      dots: $dots,
    ) {
      slides {
        id
        image
        width
        height
        headline
        subText
        buttonText
        buttonLink
        buttonColor
      }
      autoplay
      autoplaySpeed
      arrows
      dots
    }
  }
`;
export const fragments = `
  fragment SlideData on Slide {
    id
    image
    width
    height
    headline
    subText
    buttonText
    buttonLink
    buttonColor
  }
`;
export const variables = `{
  slides: getWidgetSetting("slides"),
  autoplay: getWidgetSetting("autoplay"),
  autoplaySpeed: getWidgetSetting("autoplaySpeed"),
  arrows: getWidgetSetting("arrows"),
  dots: getWidgetSetting("dots"),
}`;
