// @evershop/evershop/dist/modules/checkout/pages/admin/shippingSetting/ShippingSetting.js
import { Card } from '@components/admin/Card.js';
import { SettingMenu } from '@components/admin/SettingMenu.js';
import React from 'react';
import { Zones } from './shippingSetting/Zones.js';
export default function ShippingSetting({ createShippingZoneApi }) {
    return (React.createElement("div", { className: "main-content-inner" },
        React.createElement("div", { className: "grid grid-cols-6 gap-x-5 grid-flow-row " },
            React.createElement("div", { className: "col-span-2" },
                React.createElement(SettingMenu, null)),
            React.createElement("div", { className: "col-span-4" },
                React.createElement(Card, null,
                    React.createElement(Card.Session, { title: "Shipping" },
                        React.createElement("div", null, "Choose where you ship and how much you charge for shipping.")),
                    React.createElement(Zones, { createShippingZoneApi: createShippingZoneApi }))))));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    createShippingZoneApi: url(routeId: "createShippingZone")
  }
`;
