// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit/variants/CreateVariantGroup.js
import Spinner from '@components/admin/Spinner.js';
import React from 'react';
import { useFormContext } from 'react-hook-form';
import { toast } from 'react-toastify';
import { useQuery } from 'urql';
const AttributesQuery = `
  query Query($filters: [FilterInput]) {
    attributes(filters: $filters) {
      items {
        attributeId
        attributeCode
        attributeName
        options {
          value: attributeOptionId
          text: optionText
        }
      }
    }
  }
`;
export const CreateVariantGroup = ({ currentProductUuid, createVariantGroupApi, setGroup }) => {
    var _a, _b, _c;
    const [attributes, setAttributes] = React.useState([]);
    const { getValues } = useFormContext();
    const groupId = getValues('group_id');
    const onCreate = async (e) => {
        e.preventDefault();
        const response = await fetch(createVariantGroupApi, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                attribute_codes: attributes.map((a) => a),
                attribute_group_id: groupId
            })
        }).then((r) => r.json());
        if (!response.error) {
            // Call addItemApi to add the current product to the new variant group
            const addItemApi = response.data.addItemApi;
            const addItemResponse = await fetch(addItemApi, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    product_id: currentProductUuid
                })
            }).then((r) => r.json());
            if (addItemResponse.error) {
                toast.error(addItemResponse.error.message);
                return;
            }
            setGroup({
                variantGroupId: response.data.variant_group_id,
                addItemApi: response.data.addItemApi,
                attributes: response.data.attributes.map((attribute) => ({
                    attributeCode: attribute.attribute_code,
                    uuid: attribute.uuid,
                    attributeName: attribute.attribute_name,
                    attributeId: attribute.attribute_id,
                    options: attribute.options.map((option) => ({
                        optionId: option.attribute_option_id,
                        optionText: option.option_text
                    }))
                }))
            });
        }
        else {
            toast.error(response.error.message);
        }
    };
    const [result] = useQuery({
        query: AttributesQuery,
        variables: {
            filters: [
                { key: 'type', operation: 'eq', value: 'select' },
                { key: 'group', operation: 'eq', value: groupId }
            ]
        },
        pause: !groupId
    });
    const { data, fetching, error } = result;
    if (fetching) {
        return (React.createElement("div", { className: "flex justify-center items-center" },
            React.createElement(Spinner, { width: 30, height: 30 })));
    }
    if (error) {
        return React.createElement("p", { className: "text-critical" }, error.message);
    }
    return (React.createElement("div", null,
        React.createElement("div", null,
            (((_a = data === null || data === void 0 ? void 0 : data.attributes) === null || _a === void 0 ? void 0 : _a.items) || []).length > 0 && (React.createElement("div", null,
                React.createElement("div", null,
                    React.createElement("span", null, "Select the list of attribute")),
                (((_b = data === null || data === void 0 ? void 0 : data.attributes) === null || _b === void 0 ? void 0 : _b.items) || []).map((a) => (React.createElement("label", { key: a.attributeCode, className: "flex items-center gap-2" },
                    React.createElement("input", { type: "checkbox", onChange: (e) => {
                            if (e.target.checked) {
                                setAttributes(attributes.concat(a.attributeCode));
                            }
                            else {
                                setAttributes(attributes.filter((attr) => a.attributeCode !== attr));
                            }
                        } }),
                    React.createElement("span", null, a.attributeName)))),
                React.createElement("div", { className: "mt-5" },
                    React.createElement("a", { className: "text-interactive hover:underline", href: "#", onClick: (e) => onCreate(e) }, "Create")))),
            (((_c = data === null || data === void 0 ? void 0 : data.attributes) === null || _c === void 0 ? void 0 : _c.items) || []).length === 0 && (React.createElement("div", { className: "alert alert-danger", role: "alert" }, "There is no \"Select\" attribute available.")))));
};
