// @evershop/evershop/dist/modules/catalog/pages/admin/productEdit/variants/VariantModal.js
import { Card } from '@components/admin/Card.js';
import { ImageUploader } from '@components/admin/ImageUploader.js';
import Spinner from '@components/admin/Spinner.js';
import Button from '@components/common/Button.js';
import { InputField } from '@components/common/form/InputField.js';
import { NumberField } from '@components/common/form/NumberField.js';
import { SelectField } from '@components/common/form/SelectField.js';
import { ToggleField } from '@components/common/form/ToggleField.js';
import React, { useEffect, useState } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';
import { toast } from 'react-toastify';
import { useQuery } from 'urql';
const AttributesQuery = `
  query Query($filters: [FilterInput]) {
    attributes(filters: $filters) {
      items {
        attributeId
        attributeCode
        attributeName
        options {
          value: attributeOptionId
          label: optionText
        }
      }
    }
  }
`;
export const VariantModal = ({ variant, variantGroup, createProductApi, closeModal }) => {
    var _a, _b;
    const [saving, setSaving] = useState(false);
    const { watch, register, control } = useFormContext();
    const { fields, append, remove, replace } = useFieldArray({
        name: 'variant_images',
        control
    });
    const variantData = watch([
        'url_key',
        'variant_sku',
        'variant_visibility',
        'variant_qty',
        'variant_status',
        'variant_attributes',
        'variant_images'
    ]);
    const { getValues, trigger } = useFormContext();
    const addVariantItem = async (uuid, addVariantItemApi) => {
        const response = await fetch(addVariantItemApi, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                product_id: uuid
            })
        });
        const responseJson = await response.json();
        return responseJson;
    };
    const saveVariant = async () => {
        var _a;
        setSaving(true);
        const productFormData = getValues();
        const formData = {
            ...productFormData,
            product_id: (variant === null || variant === void 0 ? void 0 : variant.product.productId) || '',
            sku: variantData[1],
            visibility: variantData[2],
            qty: variantData[3],
            status: variantData[4],
            url_key: `${variantData[0]}-${variantData[1]}`,
            images: (variantData[6] || []).map((image) => image.url),
            attributes: ((_a = productFormData.attributes) === null || _a === void 0 ? void 0 : _a.map((attr) => {
                var _a;
                if ((_a = variantData[5]) === null || _a === void 0 ? void 0 : _a[attr.attribute_code]) {
                    return {
                        ...attr,
                        value: variantData[5][attr.attribute_code]
                    };
                }
                return attr;
            })) || []
        };
        const response = await fetch(variant ? variant.product.updateApi : createProductApi, {
            method: variant ? 'PATCH' : 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(formData)
        });
        const responseJson = await response.json();
        if (responseJson.error) {
            toast.error(responseJson.error.message);
        }
        else if (!variant) {
            const addVariantResponse = await addVariantItem(responseJson.data.uuid, variantGroup.addItemApi);
            if (addVariantResponse.error) {
                toast.error(addVariantResponse.error.message);
            }
            else {
                toast.success('Variant created successfully');
                closeModal();
            }
        }
        else {
            toast.success('Variant updated successfully');
            closeModal();
        }
        setSaving(false);
    };
    useEffect(() => {
        if (variant) {
            const images = (variant === null || variant === void 0 ? void 0 : variant.product.image)
                ? [variant.product.image].concat(variant.product.gallery || [])
                : [];
            replace(images.map((image) => ({
                uuid: image.uuid,
                path: image.path,
                url: image.url
            })));
        }
    }, []);
    const [result] = useQuery({
        query: AttributesQuery,
        variables: {
            filters: [
                {
                    key: 'code',
                    operation: 'in',
                    value: variantGroup.attributes.map((a) => a.attributeCode).join(',')
                }
            ]
        }
    });
    const { data, fetching, error } = result;
    if (fetching) {
        return (React.createElement("div", { className: "p-2 flex justify-center items-center" },
            React.createElement(Spinner, { width: 30, height: 30 })));
    }
    if (error) {
        return React.createElement("p", { className: "text-critical" }, error.message);
    }
    return (React.createElement(Card, { title: variant ? 'Edit Variant' : 'New Variant' },
        React.createElement(Card.Session, null,
            React.createElement("div", { className: "grid grid-cols-2 gap-x-5" },
                React.createElement("div", { className: "col-span-1" },
                    React.createElement(ImageUploader, { currentImages: (variant === null || variant === void 0 ? void 0 : variant.product.image)
                            ? [variant.product.image].concat(variant.product.gallery || [])
                            : [], allowDelete: true, allowSwap: true, onDelete: (image) => {
                            const index = fields.findIndex((field) => field.uuid === image.uuid);
                            if (index !== -1) {
                                remove(index);
                            }
                        }, onUpload: (images) => {
                            const newImages = images.map((image) => ({
                                id: image.uuid,
                                path: image.path,
                                url: image.url
                            }));
                            append(newImages);
                        }, onSortEnd: (oldIndex, newIndex) => {
                            const newImages = [...fields];
                            const [movedImage] = newImages.splice(oldIndex, 1);
                            newImages.splice(newIndex, 0, movedImage);
                            replace(newImages);
                        }, targetPath: `catalog/${Math.floor(Math.random() * (9999 - 1000)) + 1000}/${Math.floor(Math.random() * (9999 - 1000)) + 1000}` })),
                React.createElement("div", { className: "col-span-1" },
                    React.createElement("div", { className: "grid grid-cols-2 gap-x-2 border-b border-divider pb-4 mb-4" }, data.attributes.items.map((a) => {
                        var _a;
                        return (React.createElement("div", { key: a.attributeCode, className: "mt-2 col" },
                            React.createElement("div", null,
                                React.createElement("label", null, a.attributeName)),
                            React.createElement(SelectField, { name: `variant_attributes.${a.attributeCode}`, placeholder: "Select an option", required: true, validation: {
                                    required: 'This field is required'
                                }, defaultValue: ((_a = variant === null || variant === void 0 ? void 0 : variant.attributes.find((attr) => attr.attributeCode === a.attributeCode)) === null || _a === void 0 ? void 0 : _a.optionId.toString()) || '', options: a.options })));
                    })),
                    React.createElement("div", { className: "grid grid-cols-3 gap-x-2 border-b border-divider pb-4 mb-4" },
                        React.createElement("div", null,
                            React.createElement("div", null, "SKU"),
                            React.createElement(InputField, { name: "variant_sku", placeholder: "Enter SKU", required: true, validation: {
                                    required: 'SKU is required'
                                }, defaultValue: (_a = variant === null || variant === void 0 ? void 0 : variant.product) === null || _a === void 0 ? void 0 : _a.sku })),
                        React.createElement("div", null,
                            React.createElement("div", null, "Qty"),
                            React.createElement(NumberField, { name: "variant_qty", required: true, placeholder: "Enter quantity", validation: {
                                    required: 'Qty is required'
                                }, allowDecimals: false, defaultValue: ((_b = variant === null || variant === void 0 ? void 0 : variant.product) === null || _b === void 0 ? void 0 : _b.qty) || 0 }))),
                    React.createElement("div", { className: "grid grid-cols-3 gap-x-2" },
                        React.createElement("div", null,
                            React.createElement("div", null, "Status"),
                            React.createElement(ToggleField, { name: "variant_status", trueValue: true, falseValue: false, defaultValue: (variant === null || variant === void 0 ? void 0 : variant.product.status) === 1 })),
                        React.createElement("div", null,
                            React.createElement("div", null, "Visibility"),
                            React.createElement(ToggleField, { name: "variant_visibility", trueValue: true, falseValue: false, defaultValue: (variant === null || variant === void 0 ? void 0 : variant.product.visibility) === 1 })))))),
        React.createElement(Card.Session, null,
            React.createElement("div", { className: "flex justify-end" },
                React.createElement("div", { className: "grid grid-cols-2 gap-2" },
                    React.createElement(Button, { title: "Cancel", variant: "danger", onAction: closeModal }),
                    React.createElement(Button, { title: "Save", variant: "primary", isLoading: saving, onAction: async () => {
                            const isValid = await trigger();
                            if (isValid) {
                                await saveVariant();
                            }
                        } }))))));
};
