// @evershop/evershop/dist/modules/oms/pages/admin/dashboard/Bestsellers.js
import { Card } from '@components/admin/Card.js';
import React from 'react';
import './Bestsellers.scss';
import { Image } from '@components/common/Image.js';
import { ProductNoThumbnail } from '@components/common/ProductNoThumbnail.js';
export default function BestSellers({ bestSellers, listUrl }) {
    return (React.createElement(Card, { title: "Best Sellers", actions: [
            {
                name: 'All products',
                onAction: () => {
                    window.location.href = listUrl;
                }
            }
        ] },
        React.createElement(Card.Session, null,
            React.createElement("table", { className: "listing bestsellers" },
                React.createElement("tbody", null,
                    bestSellers.length === 0 && (React.createElement("tr", null,
                        React.createElement("td", { align: "left" }, "Look like you just started. No bestsellers yet."),
                        React.createElement("td", null, " "))),
                    bestSellers.map((p, i) => {
                        var _a, _b;
                        return (React.createElement("tr", { key: i },
                            React.createElement("td", null,
                                React.createElement("div", { className: " flex justify-left" },
                                    React.createElement("div", { className: "flex justify-start gap-2 items-center" },
                                        React.createElement("div", { className: "grid-thumbnail text-border border border-divider p-2 rounded" },
                                            ((_a = p.image) === null || _a === void 0 ? void 0 : _a.url) && (React.createElement(Image, { src: p.image.url, alt: p.name, width: 50, height: 50 })),
                                            !((_b = p.image) === null || _b === void 0 ? void 0 : _b.url) && (React.createElement(ProductNoThumbnail, { width: 50, height: 50 }))),
                                        React.createElement("div", null,
                                            React.createElement("a", { href: p.editUrl || '', className: "font-semibold hover:underline" }, p.name))))),
                            React.createElement("td", null),
                            React.createElement("td", null, p.price.regular.text),
                            React.createElement("td", null,
                                p.soldQty,
                                " sold")));
                    }))))));
}
export const layout = {
    areaId: 'leftSide',
    sortOrder: 20
};
export const query = `
  query Query {
    bestSellers {
      name
      price {
        regular {
          value
          text
        }
      }
      soldQty
      image {
        url
      }
      editUrl
    }
    listUrl: url(routeId: "productGrid")
  }
`;
