// @evershop/evershop/dist/modules/promotion/pages/admin/couponNew/CouponNewForm.js
import { Card } from '@components/admin/Card.js';
import { FormButtons } from '@components/admin/FormButtons.js';
import Area from '@components/common/Area.js';
import { Form } from '@components/common/form/Form.js';
import React from 'react';
import { toast } from 'react-toastify';
export default function CouponNewForm({ action, gridUrl }) {
    return (React.createElement(Form, { action: action, method: "POST", id: "couponNewForm", onSuccess: (response) => {
            toast.success('Coupon created successfully!');
            const editUrl = response.data.links.find((link) => link.rel === 'edit').href;
            window.location.href = editUrl;
        }, submitBtn: false },
        React.createElement("div", { className: "grid grid-cols-1 gap-5" },
            React.createElement(Card, { title: "General" },
                React.createElement(Card.Session, null,
                    React.createElement(Area, { id: "couponEditGeneral", noOuter: true }))),
            React.createElement(Card, { title: "Discount Type" },
                React.createElement(Card.Session, null,
                    React.createElement(Area, { id: "couponEditDiscountType", noOuter: true }))),
            React.createElement("div", { className: "grid grid-cols-3 gap-x-5 grid-flow-row " },
                React.createElement("div", { className: "col-span-2 grid grid-cols-1 gap-5 auto-rows-max" },
                    React.createElement(Card, { title: "Order conditions" },
                        React.createElement(Card.Session, null,
                            React.createElement(Area, { id: "couponEditLeft", noOuter: true, className: "col-8" })))),
                React.createElement("div", { className: "col-span-1 grid grid-cols-1 gap-5 auto-rows-max" },
                    React.createElement(Card, { title: "Customer conditions" },
                        React.createElement(Card.Session, null,
                            React.createElement(Area, { id: "couponEditRight", className: "col-4", noOuter: true })))))),
        React.createElement(FormButtons, { cancelUrl: gridUrl, formId: "couponNewForm" })));
}
export const layout = {
    areaId: 'content',
    sortOrder: 10
};
export const query = `
  query Query {
    action: url(routeId: "createCoupon")
    gridUrl: url(routeId: "couponGrid")
  }
`;
