// @evershop/evershop/dist/modules/checkout/pages/frontStore/checkoutSuccess/Summary.js
import React from 'react';
import './Summary.scss';
import { OrderSummaryItems } from '@components/frontStore/checkout/OrderSummaryItems.js';
import { OrderTotalSummary } from '@components/frontStore/checkout/OrderTotalSummary.js';
export default function Summary({ order }) {
    return (React.createElement("div", { className: "checkout__summary h-full hidden md:block" },
        React.createElement(OrderSummaryItems, { items: order.items }),
        React.createElement(OrderTotalSummary, { shippingCost: order.shippingFeeInclTax.text, subTotal: order.subTotal.text, total: order.grandTotal.text, shippingMethod: order.shippingMethodName, coupon: order.coupon || '', discountAmount: order.discountAmount.text, taxAmount: order.totalTaxAmount.text })));
}
export const layout = {
    areaId: 'checkoutSuccessPageRight',
    sortOrder: 10
};
export const query = `
  query Query {
    order (uuid: getContextValue('orderId')) {
      orderNumber
      discountAmount {
        text
      }
      coupon
      shippingMethodName
      shippingFeeInclTax {
        text
      }
      totalTaxAmount {
        text
      }
      subTotal {
        text
      }
      subTotalInclTax {
        text
      }
      grandTotal {
        text
      }
      items {
        uuid
        productName
        thumbnail
        productSku
        qty
        productUrl
        productPrice {
          text
        }
        productPriceInclTax {
          text
        }
        variantOptions {
          attributeCode
          attributeName
          attributeId
          optionId
          optionText
        }
        lineTotalInclTax {
          text
        }
        lineTotal {
          text
        }
      }
    }
  }
`;
